<!DOCTYPE html>

<html lang="Python" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width,initial-scale=1">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <meta name="lang:clipboard.copy" content="Copy to clipboard">
  <meta name="lang:clipboard.copied" content="Copied to clipboard">
  <meta name="lang:search.language" content="en">
  <meta name="lang:search.pipeline.stopwords" content="True">
  <meta name="lang:search.pipeline.trimmer" content="True">
  <meta name="lang:search.result.none" content="No matching documents">
  <meta name="lang:search.result.one" content="1 matching document">
  <meta name="lang:search.result.other" content="# matching documents">
  <meta name="lang:search.tokenizer" content="[\s\-]+">

  
    <link href="https://fonts.gstatic.com/" rel="preconnect" crossorigin>
    <link href="https://fonts.googleapis.com/css?family=Roboto+Mono:400,500,700|Roboto:300,400,400i,700&display=fallback" rel="stylesheet">

    <style>
      body,
      input {
        font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif
      }

      code,
      kbd,
      pre {
        font-family: "Roboto Mono", "Courier New", Courier, monospace
      }
    </style>
  

  <link rel="stylesheet" href="_static/stylesheets/application.css"/>
  <link rel="stylesheet" href="_static/stylesheets/application-palette.css"/>
  <link rel="stylesheet" href="_static/stylesheets/application-fixes.css"/>
  
  <link rel="stylesheet" href="_static/fonts/material-icons.css"/>
  
  <meta name="theme-color" content="#3f51b5">
  <script src="_static/javascripts/modernizr.js"></script>
  
  
  
    <title>SCOTCH module &#8212; SCOTCH &#39;1.1.0&#39; documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=e7352e39" />
    <link rel="stylesheet" type="text/css" href="_static/material.css?v=79c92029" />
    <script src="_static/documentation_options.js?v=0398fb1d"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="initialize module" href="initialize.html" />
    <link rel="prev" title="NMTF module" href="NMTF.html" />
  
   

  </head>
  <body dir=ltr
        data-md-color-primary=blue-grey data-md-color-accent=blue>
  
  <svg class="md-svg">
    <defs data-children-count="0">
      
      <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
      
    </defs>
  </svg>
  
  <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer">
  <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search">
  <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
  <a href="#SCOTCH" tabindex="1" class="md-skip"> Skip to content </a>
  <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex navheader">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="index.html" title="SCOTCH &#39;1.1.0&#39; documentation"
           class="md-header-nav__button md-logo">
          
            &nbsp;
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          <span class="md-header-nav__topic">SCOTCH '1.1.0' documentation</span>
          <span class="md-header-nav__topic"> SCOTCH module </span>
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
        
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" action="search.html" method="get" name="search">
      <input type="text" class="md-search__input" name="q" placeholder=""Search""
             autocapitalize="off" autocomplete="off" spellcheck="false"
             data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>

      </div>
      
      
  
  <script src="_static/javascripts/version_dropdown.js"></script>
  <script>
    var json_loc = ""versions.json"",
        target_loc = "../",
        text = "Versions";
    $( document ).ready( add_version_dropdown(json_loc, target_loc, text));
  </script>
  

    </div>
  </nav>
</header>

  
  <div class="md-container">
    
    
    
  <nav class="md-tabs" data-md-component="tabs">
    <div class="md-tabs__inner md-grid">
      <ul class="md-tabs__list">
          <li class="md-tabs__item"><a href="index.html" class="md-tabs__link">SCOTCH '1.1.0' documentation</a></li>
          <li class="md-tabs__item"><a href="modules.html" class="md-tabs__link">SCOTCH</a></li>
      </ul>
    </div>
  </nav>
    <main class="md-main">
      <div class="md-main__inner md-grid" data-md-component="container">
        
          <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="index.html" title="SCOTCH &#39;1.1.0&#39; documentation" class="md-nav__button md-logo">
      
        <img src="_static/" alt=" logo" width="48" height="48">
      
    </a>
    <a href="index.html"
       title="SCOTCH &#39;1.1.0&#39; documentation">SCOTCH '1.1.0' documentation</a>
  </label>
  

</nav>
              </div>
            </div>
          </div>
          <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                
<nav class="md-nav md-nav--secondary">
    <label class="md-nav__title" for="__toc">"Contents"</label>
  <ul class="md-nav__list" data-md-scrollfix="">
        <li class="md-nav__item"><a href="#scotch--page-root" class="md-nav__link">SCOTCH module</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH</span></code></a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.add_data_from_adata" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.add_data_from_adata()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.add_data_from_file" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.add_data_from_file()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.add_scotch_embeddings_to_adata" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.add_scotch_embeddings_to_adata()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.combined_embedding_visualization" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.combined_embedding_visualization()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.combined_enrichment_visualization" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.combined_enrichment_visualization()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.make_adata_from_scotch" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.make_adata_from_scotch()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.make_top_regulators_list" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.make_top_regulators_list()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.plot_S_matrix" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.plot_S_matrix()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.plot_cooccurrence_proportions" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.plot_cooccurrence_proportions()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.plot_element_count_heatmap" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.plot_element_count_heatmap()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.plot_reconstruction_error" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.plot_reconstruction_error()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.run_enrich_analyzer" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.run_enrich_analyzer()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.visualize_enrichment_bubbleplots" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.visualize_enrichment_bubbleplots()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.visualize_marker_gene_bubbleplot_per_cell_cluster" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.visualize_marker_gene_bubbleplot_per_cell_cluster()</span></code></a>
        </li>
        <li class="md-nav__item"><a href="#SCOTCH.SCOTCH.visualize_marker_gene_bubbleplot_per_gene_cluster" class="md-nav__link"><code class="docutils literal notranslate"><span class="pre">SCOTCH.visualize_marker_gene_bubbleplot_per_gene_cluster()</span></code></a>
        </li></ul>
            </nav>
        </li></ul>
            </nav>
        </li>
    
<li class="md-nav__item"><a class="md-nav__extra_link" href="_sources/SCOTCH.rst.txt">Show Source</a> </li>

<li id="searchbox" class="md-nav__item"></li>

  </ul>
</nav>
              </div>
            </div>
          </div>
        
        <div class="md-content">
          <article class="md-content__inner md-typeset" role="main">
            
  <section id="module-SCOTCH">
<span id="scotch-module"></span><h1 id="scotch--page-root">SCOTCH module<a class="headerlink" href="#scotch--page-root" title="Link to this heading">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">SCOTCH.</span></span><span class="sig-name descname"><span class="pre">SCOTCH</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">k1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">term_tol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_l_u</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_l_v</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_a_u</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_a_v</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_lambda</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shape_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mid_epoch_param</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">init_style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'random'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_clust</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">draw_intermediate_graph</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_intermediate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">track_objective</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kill_factors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH" title="Link to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="NMTF.html#NMTF.NMTF" title="NMTF.NMTF"><code class="xref py py-class docutils literal notranslate"><span class="pre">NMTF</span></code></a></p>
<section id="scotch-class">
<h2 id="scotch-class">SCOTCH Class<a class="headerlink" href="#scotch-class" title="Link to this heading">¶</a></h2>
<p>The <cite>SCOTCH</cite> class extends from the <cite>NMTF</cite> class. It has a specific <cite>__init__</cite> method with several input parameters. The only required inputs are <cite>k1</cite> and <cite>k2</cite>.</p>
<p><strong>__init__ Input Parameters:</strong></p>
<ul class="simple">
<li><p><strong>k1, k2</strong> (<em>int</em>):
Lower dimension size of <cite>U</cite> and <cite>V</cite>. <em>(required)</em></p></li>
<li><p><strong>verbose</strong> (<em>bool</em>, optional):
If <cite>True</cite>, prints messages. <em>(default: True)</em></p></li>
<li><p><strong>max_iter</strong> (<em>int</em>, optional):
Maximum number of iterations. <em>(default: 100)</em></p></li>
<li><p><strong>seed</strong> (<em>int</em>, optional):
Random seed for initialization. <em>(default: 1001)</em></p></li>
<li><p><strong>term_tol</strong> (<em>float</em>, optional):
Relative error threshold for convergence. <em>(default: 1e-5)</em></p></li>
<li><p><strong>max_l_u</strong> (<em>float</em>, optional):
Maximum regularization on <cite>U</cite>. <em>(default: 0)</em></p></li>
<li><p><strong>max_l_v</strong> (<em>float</em>, optional):
Maximum regularization on <cite>V</cite>. <em>(default: 0)</em></p></li>
<li><p><strong>max_a_u</strong> (<em>float</em>, optional):
Maximum sparse regularization on <cite>U</cite>. <em>(default: 0, change at own risk)</em></p></li>
<li><p><strong>max_a_v</strong> (<em>float</em>, optional):
Maximum sparse regularization on <cite>V</cite>. <em>(default: 0, change at own risk)</em></p></li>
<li><p><strong>var_lambda</strong> (<em>bool</em>, optional):
If <cite>True</cite>, the regularization parameters <cite>l_U</cite> and <cite>l_V</cite> increase to max value using a sigmoid scheduler. Generally set to <cite>False</cite>. <em>(default: False)</em></p></li>
<li><p><strong>var_alpha</strong> (<em>bool</em>, optional):
If <cite>True</cite>, the regularization parameters <cite>a_U</cite> and <cite>a_V</cite> increase to max value using a sigmoid scheduler. Generally set to <cite>False</cite>. <em>(default: False)</em></p></li>
<li><p><strong>shape_param</strong> (<em>float</em>, optional):
Controls the rate of increase for <cite>l_U</cite>, <cite>l_V</cite>, <cite>a_U</cite>, and <cite>a_V</cite> when <cite>var_lambda=True</cite>. <em>(default: 10)</em></p></li>
<li><p><strong>mid_epoch_param</strong> (<em>int</em>, optional):
Sets the epoch where <cite>l_U</cite>, <cite>l_V</cite>, <cite>a_U</cite>, and <cite>a_V</cite> reach half of their max values if <cite>var_lambda=True</cite>. <em>(default: 5)</em></p></li>
<li><p><strong>init_style</strong> (<em>str</em>, optional):
Initialization method for SCOTCH. Should be either <cite>"random"</cite> or <cite>"nnsvd"</cite>. <em>(default: "random")</em></p></li>
<li><p><strong>save_clust</strong> (<em>bool</em>, optional):
Whether to save cluster assignments after each epoch. <em>(default: False)</em></p></li>
<li><p><strong>draw_intermediate_graph</strong> (<em>bool</em>, optional):
If <cite>True</cite>, draws and saves the matrix representation after each epoch. These can be saved as a GIF. <em>(default: False)</em></p></li>
<li><p><strong>track_objective</strong> (<em>bool</em>, deprecated):
<em>(default: False)</em></p></li>
<li><p><strong>kill_factors</strong> (<em>bool</em>, optional):
If <cite>True</cite>, SCOTCH will halt updates if any factors in <cite>U</cite> and <cite>V</cite> reach zero. <em>(default: False)</em></p></li>
<li><p><strong>device</strong> (<em>str</em>, optional):
Specifies the device to run SCOTCH on: <cite>"cpu"</cite> or <cite>"cuda:"</cite>. <em>(default: "cpu")</em></p></li>
<li><p><strong>out_path</strong> (<em>str</em>, optional):
Directory to save SCOTCH output files. <em>(default: '.')</em></p></li>
</ul>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.add_data_from_adata">
<span class="sig-name descname"><span class="pre">add_data_from_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.add_data_from_adata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.add_data_from_adata" title="Link to this definition">¶</a></dt>
<dd><p>Loads data from AnnData object into SCOTCH framework.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- anndata.AnnData object to extract data from. Transforms adata.X to PyTorch object.</p>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.add_data_from_file">
<span class="sig-name descname"><span class="pre">add_data_from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.add_data_from_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.add_data_from_file" title="Link to this definition">¶</a></dt>
<dd><p>Loads matrix representation into PyTorch tensor object to run with SCOTCH.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file</strong> (<em>str</em>) -- The file path to load data from and should have the valid extensions like '.pt', '.txt', or '.h5ad'.</p>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.add_scotch_embeddings_to_adata">
<span class="sig-name descname"><span class="pre">add_scotch_embeddings_to_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.add_scotch_embeddings_to_adata"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.add_scotch_embeddings_to_adata" title="Link to this definition">¶</a></dt>
<dd><p>Adds SCOTCH objects to an AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prefix</strong> (<em>str</em>) -- Prefix to add to AnnData objects created by SCOTCH.</p></li>
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- The AnnData object to which SCOTCH embeddings will be added.</p></li>
</ul>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.combined_embedding_visualization">
<span class="sig-name descname"><span class="pre">combined_embedding_visualization</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_cluster_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_embedding_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_embedding'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_point_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sample'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">S_matrix_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'S_matrix'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.combined_embedding_visualization"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.combined_embedding_visualization" title="Link to this definition">¶</a></dt>
<dd><p>Generate a combined visualization of embeddings from the data, with options to color by metadata.</p>
<p>This function is designed to visualize embeddings (e.g., UMAP, PCA, t-SNE) stored in <cite>adata</cite>, optionally
allowing users to color the points by specific metadata columns (like cell type or condition). It arranges
one or more plots in a grid layout.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- An AnnData object containing single-cell data. Must contain embeddings in <cite>adata.obsm</cite>.</p></li>
<li><p><strong>gene_embedding_id</strong> (<em>str</em><em>, </em><em>optional</em>) -- The key in <cite>adata.obsm</cite> where the embedding is stored (e.g., <cite>'X_umap'</cite> for UMAP). Default is <cite>'gene_embedding'</cite>.</p></li>
<li><p><strong>top_k</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default=5</em><em>)</em>) -- the number of top features to display per gene cluster. Must be a positive integer.</p></li>
<li><p><strong>max_point_size</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default=2</em><em>)</em>) -- Max point size for V bubble plot.</p></li>
<li><p><strong>palette</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default='viridis'</em><em>)</em>) -- The color palette to use for the scatter plot when coloring points.</p></li>
<li><p><strong>var1</strong> (<em>str</em><em>, </em><em>optional</em>) -- The primary variable (from <cite>adata.obs</cite>) to use for heatmap plotting (e.g., 'cell_clusters').</p></li>
<li><p><strong>var2</strong> (<em>str</em><em>, </em><em>optional</em>) -- The secondary variable (from <cite>adata.obs</cite>) for co-occurrence and proportions (e.g., 'sample').</p></li>
<li><p><strong>S_matrix_id</strong> (<em>str</em><em>, </em><em>optional</em>) -- <p>Key in <cite>adata.uns</cite> for an externally referenced matrix (e.g., factor matrix or count matrix).</p>
<p>Returns</p>
</p></li>
<li><p><strong>-------</strong></p></li>
<li><p><strong>matplotlib.figure.Figure</strong> -- A matplotlib figure containing the generated visualizations.</p></li>
</ul>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.combined_enrichment_visualization">
<span class="sig-name descname"><span class="pre">combined_enrichment_visualization</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enrich_object_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_point_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sample'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">S_matrix_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.combined_enrichment_visualization"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.combined_enrichment_visualization" title="Link to this definition">¶</a></dt>
<dd><p>Create a 2x3 subplot visualization combining enrichment bubble plots, element count heatmaps,
co-occurrence proportions, and S matrix heatmaps.</p>
<p>This visualization provides insights into cellular data enrichment and relationships between variables.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- An AnnData object containing single-cell data. Must contain <cite>obs</cite>, <cite>uns</cite>, and required data matrices.</p></li>
<li><p><strong>enrich_object_id</strong> (<em>str</em>) -- The identifier for enrichment data in <cite>adata.uns</cite>.</p></li>
<li><p><strong>top_k</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default=5</em><em>)</em>) -- The number of top enrichment terms to display in visualizations.
Must be a positive integer.</p></li>
<li><p><strong>max_point_size</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default=100</em><em>)</em>) -- The maximum size for the points in the bubble plot. Determines the largest bubble size.</p></li>
<li><p><strong>palette</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default='viridis'</em><em>)</em>) -- The color palette used for the bubble plot.</p></li>
<li><p><strong>var1</strong> (<em>str</em><em>, </em><em>optional</em>) -- The first variable for co-occurrence proportions. Must exist in <cite>adata.obs.columns</cite>.</p></li>
<li><p><strong>var2</strong> (<em>str</em><em>, </em><em>optional</em>) -- The second variable for co-occurrence proportions. Must exist in <cite>adata.obs.columns</cite>.</p></li>
<li><p><strong>S_matrix_id</strong> (<em>str</em><em>, </em><em>optional</em>) -- The key in <cite>adata.uns</cite> for the additional heatmap matrix. The associated value must be a 2D matrix.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A matplotlib figure object containing the generated subplots.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.make_adata_from_scotch">
<span class="sig-name descname"><span class="pre">make_adata_from_scotch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.make_adata_from_scotch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.make_adata_from_scotch" title="Link to this definition">¶</a></dt>
<dd><p>Create an AnnData object from the given data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) -- The instance of the class containing the data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) -- A string appended to the generated AnnData objects.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An AnnData object containing the processed data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>anndata.AnnData</p>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.make_top_regulators_list">
<span class="sig-name descname"><span class="pre">make_top_regulators_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_cluster_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_embedding_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_embedding'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.make_top_regulators_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.make_top_regulators_list" title="Link to this definition">¶</a></dt>
<dd><p>Create a list of top regulators for each gene cluster based on gene embeddings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) -- The instance of the class containing the data.</p></li>
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- An AnnData object containing single-cell gene expression data.</p></li>
<li><p><strong>gene_cluster_id</strong> (<em>str</em>) -- The key for the gene clusters stored in <cite>adata.var</cite>. (default is "gene_clusters")</p></li>
<li><p><strong>gene_embedding_id</strong> (<em>str</em>) -- The key for the gene embedding matrix stored in <cite>adata.varm</cite>. (default is "gene_embedding")</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) -- The string utilized when adding SCOTCH data to anndata. Use instead of gene_cluster_id and gene_embedding_id. (default is None)</p></li>
<li><p><strong>top_k</strong> (<em>int</em><em>, </em><em>optional</em>) -- The number of top genes to select per cluster (default is 5).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of tuples, each containing the cluster index and the top <cite>top_k</cite> genes for that cluster.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list of tuples</p>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.plot_S_matrix">
<span class="sig-name descname"><span class="pre">plot_S_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">S_matrix_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'S_matrix'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.plot_S_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.plot_S_matrix" title="Link to this definition">¶</a></dt>
<dd><p>Plot a heatmap of the S matrix stored in <cite>adata.uns</cite> under the given key (<cite>S_matrix_id</cite>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- An AnnData object containing the single-cell data.</p></li>
<li><p><strong>S_matrix_id</strong> (<em>str</em>) -- The key in <cite>adata.uns</cite> where the S matrix is stored. Default is 'S_matrix'.</p></li>
<li><p><strong>palette</strong> (<em>str</em>) -- The color palette to use for the heatmap. Default is 'viridis'.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) -- A matplotlib Axes object to plot the heatmap on. If not provided, a new figure and Axes will be created.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A matplotlib Figure object containing the heatmap, or None if <cite>ax</cite> is provided.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure or None</p>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.plot_cooccurrence_proportions">
<span class="sig-name descname"><span class="pre">plot_cooccurrence_proportions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">field_1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">field_2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sample'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Reds'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.plot_cooccurrence_proportions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.plot_cooccurrence_proportions" title="Link to this definition">¶</a></dt>
<dd><p>Generate a heatmap of co-occurrence proportions between two categorical variables.</p>
<p>This function creates a heatmap to visualize the co-occurrence proportions of two categorical variables
stored in the <cite>adata.obs</cite> dataframe. The rows correspond to values from <cite>field_1</cite> and the
columns correspond to values from <cite>field_2</cite>. The heatmap displays normalized proportions per row.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- An AnnData object containing the single-cell data.
It must include <cite>adata.obs</cite> with <cite>field_1</cite> and <cite>field_2</cite> as categorical variables.</p></li>
<li><p><strong>field_1</strong> (<em>str</em><em>, </em><em>optional</em>) -- The name of the first categorical variable in <cite>adata.obs</cite> (used for heatmap rows).
Its values will define the rows in the heatmap. (default is 'cell_clusters')</p></li>
<li><p><strong>field_2</strong> (<em>str</em><em>, </em><em>optional</em>) -- The name of the second categorical variable in <cite>adata.obs</cite> (used for heatmap columns).
Its values will define the columns in the heatmap. (Default is 'sample'.)</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em>, </em><em>optional</em>) -- The color map used for the heatmap. It must be a valid Matplotlib colormap, with "Reds"
as the default. This determines the gradient colors representing value intensity in the heatmap.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em> or </em><em>None</em><em>, </em><em>optional</em>) -- A Matplotlib <cite>Axes</cite> object on which the heatmap will be plotted. If <cite>None</cite>, a new figure
and axis are created, and the function returns the generated figure. If provided, the heatmap
is plotted on the existing axis, and no figure is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A Matplotlib <cite>Figure</cite> object containing the heatmap of co-occurrence proportions,
if a new figure is generated. If <cite>ax</cite> is provided, the function returns None.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure or None</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>TypeError</strong> -- If <cite>adata</cite> is not an <cite>AnnData</cite> object, or <cite>field_1</cite> and <cite>field_2</cite> are not strings.</p></li>
<li><p><strong>ValueError</strong> -- If <cite>field_1</cite> or <cite>field_2</cite> are not found in <cite>adata.obs.columns</cite>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.plot_element_count_heatmap">
<span class="sig-name descname"><span class="pre">plot_element_count_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'vertical'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Blues'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.plot_element_count_heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.plot_element_count_heatmap" title="Link to this definition">¶</a></dt>
<dd><p>This function produces a heatmap displaying the count of unique elements in a specific column of an <cite>AnnData</cite> object.
The orientation of the heatmap can be controlled, along with customization options like the color map and axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- An AnnData object containing the single-cell data.
This parameter stores observations and variables, including metadata used for the analysis.</p></li>
<li><p><strong>field</strong> (<em>str optional.</em>) -- The column in <cite>adata.obs</cite> for which the counts should be calculated.
The unique values in this column are counted and visualized in the heatmap. Default is cell_clusters</p></li>
<li><p><strong>orientation</strong> (<em>str</em><em>, </em><em>optional</em>) -- The orientation of the heatmap (either rows or columns represent the elements being counted).
Acceptable values are 'vertical' (default) or 'horizontal'.</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em>, </em><em>optional</em>) -- The color map used to style the heatmap. For example, use 'Blues' for a blue shade gradient.
Defaults to 'Blues'.</p></li>
<li><p><strong>v_min</strong> (<em>int</em><em>, </em><em>optional</em>) -- The minimum value for the heatmap color scale.
This is useful to set a threshold for visualization. Default is 0.</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) -- A matplotlib <cite>Axes</cite> object onto which the heatmap will be drawn.
If not provided, a new figure and axes will be created.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A matplotlib <cite>Figure</cite> object containing the heatmap.
If <cite>ax</cite> is provided, then the returned <cite>Figure</cite> will be <cite>None</cite>, as the plot will be drawn on the given <cite>Axes</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure or None</p>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.plot_reconstruction_error">
<span class="sig-name descname"><span class="pre">plot_reconstruction_error</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.plot_reconstruction_error"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.plot_reconstruction_error" title="Link to this definition">¶</a></dt>
<dd></dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.run_enrich_analyzer">
<span class="sig-name descname"><span class="pre">run_enrich_analyzer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_cluster_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_regnet_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'persg'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GO'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.run_enrich_analyzer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.run_enrich_analyzer" title="Link to this definition">¶</a></dt>
<dd><p>Perform gene ontology (GO) enrichment analysis and store results in the AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) -- The instance of the class containing the data.</p></li>
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- An AnnData object containing single-cell gene expression data.</p></li>
<li><p><strong>gene_cluster_id</strong> (<em>str</em>) -- Identifier for the gene cluster to be analyzed.</p></li>
<li><p><strong>go_regnet_file</strong> (<em>str</em>) -- File path to the gene ontology (GO) enrichment file.</p></li>
<li><p><strong>fdr</strong> (<em>float</em><em>, </em><em>optional</em>) -- The false discovery rate threshold (default is 0.05).</p></li>
<li><p><strong>test_type</strong> (<em>str</em><em>, </em><em>optional</em>) -- Type of statistical test to be performed (default is 'persg'). Valid options are 'persg' and 'fullgraph'.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) -- Prefix for storing enrichment results in the AnnData object (default is 'GO').</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None. Enrichment results are stored in <cite>adata.uns[prefix + "enrichment"]</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.visualize_enrichment_bubbleplots">
<span class="sig-name descname"><span class="pre">visualize_enrichment_bubbleplots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enrich_object_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_cluster_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene</span> <span class="pre">cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">term_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'TermName'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">FC_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Foldenr'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">q_val_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CorrPval'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_point_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_cluster_set</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.visualize_enrichment_bubbleplots"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.visualize_enrichment_bubbleplots" title="Link to this definition">¶</a></dt>
<dd><p>Visualize enrichment results as a bubble plot, where the size of the bubbles represents log2 fold change
and the color represents -log10 p-values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) -- The instance of the class containing the data.</p></li>
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- An AnnData object containing single-cell gene expression data.</p></li>
<li><p><strong>enrich_object_id</strong> (<em>str</em>) -- The key in <cite>adata.uns</cite> containing the enrichment data.</p></li>
<li><p><strong>gene_cluster_id</strong> (<em>str</em><em>, </em><em>optional</em>) -- The column name representing gene clusters in the enrichment data (default is 'gene cluster').</p></li>
<li><p><strong>term_id</strong> (<em>str</em><em>, </em><em>optional</em>) -- The column name representing terms in the enrichment data (default is 'TermName').</p></li>
<li><p><strong>FC_id</strong> (<em>str</em><em>, </em><em>optional</em>) -- The column name representing fold change values in the enrichment data (default is 'Foldenr').</p></li>
<li><p><strong>q_val_id</strong> (<em>str</em><em>, </em><em>optional</em>) -- The column name representing the corrected p-values in the enrichment data (default is 'CorrPval').</p></li>
<li><p><strong>top_k</strong> (<em>int</em><em>, </em><em>optional</em>) -- The number of top terms to select per gene cluster (default is 5).</p></li>
<li><p><strong>max_point_size</strong> (<em>int</em><em>, </em><em>optional</em>) -- The maximum size of the bubbles in the plot (default is 100).</p></li>
<li><p><strong>palette</strong> (<em>str</em><em>, </em><em>optional</em>) -- The color palette used for the plot (default is 'viridis').</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <cite>matplotlib.figure.Figure</cite> object containing the bubble plot.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.visualize_marker_gene_bubbleplot_per_cell_cluster">
<span class="sig-name descname"><span class="pre">visualize_marker_gene_bubbleplot_per_cell_cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_cluster_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell_clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_cluster_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_embedding_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_embedding'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_point_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.visualize_marker_gene_bubbleplot_per_cell_cluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.visualize_marker_gene_bubbleplot_per_cell_cluster" title="Link to this definition">¶</a></dt>
<dd><p>Visualize marker gene expression as a bubble plot, where the size of the bubbles represents the
percent of non-zero counts and the color represents the mean marker expression. Top_k are selected for each
gene cluster. Genes names are follows by the gene cluster that each gene corresponds to.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) -- The instance of the class containing the data.</p></li>
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- An AnnData object containing single-cell gene expression data.</p></li>
<li><p><strong>cell_cluster_id</strong> (<em>str</em>) -- The column name representing cell clusters in <cite>adata.obs</cite>. (default is 'cell_clusters')</p></li>
<li><p><strong>gene_cluster_id</strong> (<em>str</em>) -- The column name representing gene clusters in <cite>adata.var</cite>. (default is 'gene_clusters')</p></li>
<li><p><strong>gene_embedding_id</strong> (<em>str</em>) -- The identifier for the gene embedding used for selecting top markers. (default is 'gene_embedding')</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) -- The string utilized when adding SCOTCH data to anndata. Use instead of gene_cluster_id and gene_embedding_id. (default is None)</p></li>
<li><p><strong>top_k</strong> (<em>int</em><em>, </em><em>optional</em>) -- The number of top markers to consider per gene cluster (default is 5).</p></li>
<li><p><strong>max_point_size</strong> (<em>int</em><em>, </em><em>optional</em>) -- The maximum size of the bubbles in the plot (default is 300).</p></li>
<li><p><strong>palette</strong> (<em>str</em><em>, </em><em>optional</em>) -- The color palette used for the plot (default is 'viridis').</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) -- The matplotlib.axes.axes object to plot in. If none, new figure is generated and returned (default is None)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <cite>matplotlib.figure.Figure</cite> object containing the bubble plot if ax not passed. Else returns none.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure or None</p>
</dd>
</dl>
</dd></dl>
<dl class="py method">
<dt class="sig sig-object py" id="SCOTCH.SCOTCH.visualize_marker_gene_bubbleplot_per_gene_cluster">
<span class="sig-name descname"><span class="pre">visualize_marker_gene_bubbleplot_per_gene_cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_cluster_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_clusters'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_embedding_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_embedding'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_point_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">300</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/SCOTCH.html#SCOTCH.visualize_marker_gene_bubbleplot_per_gene_cluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#SCOTCH.SCOTCH.visualize_marker_gene_bubbleplot_per_gene_cluster" title="Link to this definition">¶</a></dt>
<dd><p>Visualize marker gene expression as a bubble plot, where the size of the bubbles represents the
percent of non-zero counts and the color represents the mean marker expression. Top_k are selected for each
gene cluster. Genes names are follows by the gene cluster that each gene corresponds to.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> (<em>object</em>) -- The instance of the class containing the data.</p></li>
<li><p><strong>adata</strong> (<em>anndata.AnnData</em>) -- An AnnData object containing single-cell gene expression data.</p></li>
<li><p><strong>gene_cluster_id</strong> (<em>str</em><em>, </em><em>optional</em>) -- The identifier for the gene cluster used for selecting top markers.</p></li>
<li><p><strong>gene_embedding_id</strong> (<em>str</em><em>, </em><em>optional</em>) -- The identifier for the gene embedding used for selecting top markers.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) -- The prefix string used when adding SCOTCH data to anndata object.</p></li>
<li><p><strong>top_k</strong> (<em>int</em><em>, </em><em>optional</em>) -- The number of top markers to consider per gene cluster (default is 5).</p></li>
<li><p><strong>max_point_size</strong> (<em>int</em><em>, </em><em>optional</em>) -- The maximum size of the bubbles in the plot (default is 300).</p></li>
<li><p><strong>palette</strong> (<em>str</em><em>, </em><em>optional</em>) -- The color palette used for the plot (default is 'viridis').</p></li>
<li><p><strong>ax</strong> (<em>matplotlib.axes.Axes</em><em>, </em><em>optional</em>) -- A <cite>matplotlib.axes.Axes</cite> object to plot the bubble plot. If not provided a new figures is generated</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A <cite>matplotlib.figure.Figure</cite> object containing the bubble plot.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>
</section>
</dd></dl>
</section>


          </article>
        </div>
      </div>
    </main>
  </div>
  <footer class="md-footer">
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
          
            <a href="NMTF.html" title="NMTF module"
               class="md-flex md-footer-nav__link md-footer-nav__link--prev"
               rel="prev">
              <div class="md-flex__cell md-flex__cell--shrink">
                <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
              </div>
              <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
                <span class="md-flex__ellipsis">
                  <span
                      class="md-footer-nav__direction"> "Previous" </span> NMTF module </span>
              </div>
            </a>
          
          
            <a href="initialize.html" title="initialize module"
               class="md-flex md-footer-nav__link md-footer-nav__link--next"
               rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title"><span
                class="md-flex__ellipsis"> <span
                class="md-footer-nav__direction"> "Next" </span> initialize module </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink"><i
                class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          
        </a>
        
      </nav>
    </div>
    <div class="md-footer-meta md-typeset">
      <div class="md-footer-meta__inner md-grid">
        <div class="md-footer-copyright">
          <div class="md-footer-copyright__highlight">
              &#169; Copyright 2025, Spencer Halberg-Spencer, Harmon Bhasin, Sushmita Roy.
              
          </div>
            Created using
            <a href="http://www.sphinx-doc.org/">Sphinx</a> 7.3.7.
             and
            <a href="https://github.com/bashtage/sphinx-material/">Material for
              Sphinx</a>
        </div>
      </div>
    </div>
  </footer>
  <script src="_static/javascripts/application.js"></script>
  <script>app.initialize({version: "1.0.4", url: {base: ".."}})</script>
  </body>
</html>