<!DOCTYPE html>

<html lang="Python" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <meta name="viewport" content="width=device-width,initial-scale=1">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <meta name="lang:clipboard.copy" content="Copy to clipboard">
  <meta name="lang:clipboard.copied" content="Copied to clipboard">
  <meta name="lang:search.language" content="en">
  <meta name="lang:search.pipeline.stopwords" content="True">
  <meta name="lang:search.pipeline.trimmer" content="True">
  <meta name="lang:search.result.none" content="No matching documents">
  <meta name="lang:search.result.one" content="1 matching document">
  <meta name="lang:search.result.other" content="# matching documents">
  <meta name="lang:search.tokenizer" content="[\s\-]+">

  
    <link href="https://fonts.gstatic.com/" rel="preconnect" crossorigin>
    <link href="https://fonts.googleapis.com/css?family=Roboto+Mono:400,500,700|Roboto:300,400,400i,700&display=fallback" rel="stylesheet">

    <style>
      body,
      input {
        font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif
      }

      code,
      kbd,
      pre {
        font-family: "Roboto Mono", "Courier New", Courier, monospace
      }
    </style>
  

  <link rel="stylesheet" href="../_static/stylesheets/application.css"/>
  <link rel="stylesheet" href="../_static/stylesheets/application-palette.css"/>
  <link rel="stylesheet" href="../_static/stylesheets/application-fixes.css"/>
  
  <link rel="stylesheet" href="../_static/fonts/material-icons.css"/>
  
  <meta name="theme-color" content="#3f51b5">
  <script src="../_static/javascripts/modernizr.js"></script>
  
  
  
    <title>SCOTCH &#8212; SCOTCH &#39;1.1.0&#39; documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=e7352e39" />
    <link rel="stylesheet" type="text/css" href="../_static/material.css?v=79c92029" />
    <script src="../_static/documentation_options.js?v=0398fb1d"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  
   

  </head>
  <body dir=ltr
        data-md-color-primary=blue-grey data-md-color-accent=blue>
  
  <svg class="md-svg">
    <defs data-children-count="0">
      
      <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
      
    </defs>
  </svg>
  
  <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer">
  <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search">
  <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
  <a href="#_modules/SCOTCH" tabindex="1" class="md-skip"> Skip to content </a>
  <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex navheader">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../index.html" title="SCOTCH &#39;1.1.0&#39; documentation"
           class="md-header-nav__button md-logo">
          
            &nbsp;
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          <span class="md-header-nav__topic">SCOTCH '1.1.0' documentation</span>
          <span class="md-header-nav__topic"> SCOTCH </span>
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
        
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" action="../search.html" method="get" name="search">
      <input type="text" class="md-search__input" name="q" placeholder=""Search""
             autocapitalize="off" autocomplete="off" spellcheck="false"
             data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>

      </div>
      
      
  
  <script src="../_static/javascripts/version_dropdown.js"></script>
  <script>
    var json_loc = "../"versions.json"",
        target_loc = "../../",
        text = "Versions";
    $( document ).ready( add_version_dropdown(json_loc, target_loc, text));
  </script>
  

    </div>
  </nav>
</header>

  
  <div class="md-container">
    
    
    
  <nav class="md-tabs" data-md-component="tabs">
    <div class="md-tabs__inner md-grid">
      <ul class="md-tabs__list">
          <li class="md-tabs__item"><a href="../index.html" class="md-tabs__link">SCOTCH '1.1.0' documentation</a></li>
          <li class="md-tabs__item"><a href="index.html" class="md-tabs__link">Module code</a></li>
      </ul>
    </div>
  </nav>
    <main class="md-main">
      <div class="md-main__inner md-grid" data-md-component="container">
        
          <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="../index.html" title="SCOTCH &#39;1.1.0&#39; documentation" class="md-nav__button md-logo">
      
        <img src="../_static/" alt=" logo" width="48" height="48">
      
    </a>
    <a href="../index.html"
       title="SCOTCH &#39;1.1.0&#39; documentation">SCOTCH '1.1.0' documentation</a>
  </label>
  

</nav>
              </div>
            </div>
          </div>
          <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                
<nav class="md-nav md-nav--secondary">
  <ul class="md-nav__list" data-md-scrollfix="">
    

<li id="searchbox" class="md-nav__item"></li>

  </ul>
</nav>
              </div>
            </div>
          </div>
        
        <div class="md-content">
          <article class="md-content__inner md-typeset" role="main">
            
  <h1 id="modules-scotch--page-root">Source code for SCOTCH</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">DataLoader</span> <span class="kn">import</span> <span class="n">DataLoader</span>
<span class="kn">from</span> <span class="nn">NMTF</span> <span class="kn">import</span> <span class="n">NMTF</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">issparse</span>
<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib.gridspec</span> <span class="k">as</span> <span class="nn">gridspec</span>

<span class="kn">import</span> <span class="nn">pyEnrichAnalyzer</span>


<div class="viewcode-block" id="SCOTCH">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH">[docs]</a>
<span class="k">class</span> <span class="nc">SCOTCH</span><span class="p">(</span><span class="n">NMTF</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""</span>
<span class="sd">SCOTCH Class</span>
<span class="sd">============</span>

<span class="sd">The `SCOTCH` class extends from the `NMTF` class. It has a specific `__init__` method with several input parameters. The only required inputs are `k1` and `k2`.</span>

<span class="sd">**__init__ Input Parameters:**</span>

<span class="sd">- **k1, k2** (*int*):</span>
<span class="sd">  Lower dimension size of `U` and `V`. *(required)*</span>

<span class="sd">- **verbose** (*bool*, optional):</span>
<span class="sd">  If `True`, prints messages. *(default: True)*</span>

<span class="sd">- **max_iter** (*int*, optional):</span>
<span class="sd">  Maximum number of iterations. *(default: 100)*</span>

<span class="sd">- **seed** (*int*, optional):</span>
<span class="sd">  Random seed for initialization. *(default: 1001)*</span>

<span class="sd">- **term_tol** (*float*, optional):</span>
<span class="sd">  Relative error threshold for convergence. *(default: 1e-5)*</span>

<span class="sd">- **max_l_u** (*float*, optional):</span>
<span class="sd">  Maximum regularization on `U`. *(default: 0)*</span>

<span class="sd">- **max_l_v** (*float*, optional):</span>
<span class="sd">  Maximum regularization on `V`. *(default: 0)*</span>

<span class="sd">- **max_a_u** (*float*, optional):</span>
<span class="sd">  Maximum sparse regularization on `U`. *(default: 0, change at own risk)*</span>

<span class="sd">- **max_a_v** (*float*, optional):</span>
<span class="sd">  Maximum sparse regularization on `V`. *(default: 0, change at own risk)*</span>

<span class="sd">- **var_lambda** (*bool*, optional):</span>
<span class="sd">  If `True`, the regularization parameters `l_U` and `l_V` increase to max value using a sigmoid scheduler. Generally set to `False`. *(default: False)*</span>

<span class="sd">- **var_alpha** (*bool*, optional):</span>
<span class="sd">  If `True`, the regularization parameters `a_U` and `a_V` increase to max value using a sigmoid scheduler. Generally set to `False`. *(default: False)*</span>

<span class="sd">- **shape_param** (*float*, optional):</span>
<span class="sd">  Controls the rate of increase for `l_U`, `l_V`, `a_U`, and `a_V` when `var_lambda=True`. *(default: 10)*</span>

<span class="sd">- **mid_epoch_param** (*int*, optional):</span>
<span class="sd">  Sets the epoch where `l_U`, `l_V`, `a_U`, and `a_V` reach half of their max values if `var_lambda=True`. *(default: 5)*</span>

<span class="sd">- **init_style** (*str*, optional):</span>
<span class="sd">  Initialization method for SCOTCH. Should be either `"random"` or `"nnsvd"`. *(default: "random")*</span>

<span class="sd">- **save_clust** (*bool*, optional):</span>
<span class="sd">  Whether to save cluster assignments after each epoch. *(default: False)*</span>

<span class="sd">- **draw_intermediate_graph** (*bool*, optional):</span>
<span class="sd">  If `True`, draws and saves the matrix representation after each epoch. These can be saved as a GIF. *(default: False)*</span>

<span class="sd">- **track_objective** (*bool*, deprecated):</span>
<span class="sd">  *(default: False)*</span>

<span class="sd">- **kill_factors** (*bool*, optional):</span>
<span class="sd">  If `True`, SCOTCH will halt updates if any factors in `U` and `V` reach zero. *(default: False)*</span>

<span class="sd">- **device** (*str*, optional):</span>
<span class="sd">  Specifies the device to run SCOTCH on: `"cpu"` or `"cuda:"`. *(default: "cpu")*</span>

<span class="sd">- **out_path** (*str*, optional):</span>
<span class="sd">  Directory to save SCOTCH output files. *(default: '.')*</span>
<span class="sd">"""</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">k1</span><span class="p">,</span> <span class="n">k2</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">max_iter</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">1001</span><span class="p">,</span> <span class="n">term_tol</span><span class="o">=</span><span class="mf">1e-5</span><span class="p">,</span>
                 <span class="n">max_l_u</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_l_v</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_a_u</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_a_v</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">var_lambda</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">var_alpha</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">shape_param</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">mid_epoch_param</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                 <span class="n">init_style</span><span class="o">=</span><span class="s2">"random"</span><span class="p">,</span> <span class="n">save_clust</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">draw_intermediate_graph</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">save_intermediate</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">track_objective</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">kill_factors</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="s2">"cpu"</span><span class="p">,</span> <span class="n">out_path</span><span class="o">=</span><span class="s1">'.'</span><span class="p">):</span>

        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">verbose</span><span class="p">,</span> <span class="n">max_iter</span><span class="p">,</span> <span class="n">seed</span><span class="p">,</span> <span class="n">term_tol</span><span class="p">,</span> <span class="n">max_l_u</span><span class="p">,</span> <span class="n">max_l_v</span><span class="p">,</span> <span class="n">max_a_u</span><span class="p">,</span> <span class="n">max_a_v</span><span class="p">,</span> <span class="n">k1</span><span class="p">,</span> <span class="n">k2</span><span class="p">,</span> <span class="n">var_lambda</span><span class="p">,</span>
                         <span class="n">var_alpha</span><span class="p">,</span> <span class="n">shape_param</span><span class="p">,</span> <span class="n">mid_epoch_param</span><span class="p">,</span> <span class="n">init_style</span><span class="p">,</span> <span class="n">save_clust</span><span class="p">,</span> <span class="n">draw_intermediate_graph</span><span class="p">,</span>
                         <span class="n">save_intermediate</span><span class="p">,</span> <span class="n">track_objective</span><span class="p">,</span> <span class="n">kill_factors</span><span class="p">,</span> <span class="n">device</span><span class="p">,</span> <span class="n">out_path</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">DataLoader</span> <span class="o">=</span> <span class="n">DataLoader</span><span class="p">(</span><span class="n">verbose</span><span class="p">)</span>

<div class="viewcode-block" id="SCOTCH.add_data_from_file">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.add_data_from_file">[docs]</a>
    <span class="k">def</span> <span class="nf">add_data_from_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Loads matrix representation into PyTorch tensor object to run with SCOTCH.</span>

<span class="sd">        :param file: The file path to load data from and should have the valid extensions like '.pt', '.txt', or '.h5ad'.</span>
<span class="sd">        :type file: str</span>
<span class="sd">        """</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s1">'file must be a string'</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">'The file does not exist'</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">R_OK</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">'The file is not readable'</span><span class="p">)</span>

        <span class="n">shape</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">file_extension</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">file_extension</span> <span class="o">==</span> <span class="s1">'.pt'</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">shape</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DataLoader</span><span class="o">.</span><span class="n">from_pt</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">file_extension</span> <span class="o">==</span> <span class="s1">'.txt'</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">shape</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DataLoader</span><span class="o">.</span><span class="n">from_text</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">file_extension</span> <span class="o">==</span> <span class="s1">'.h5ad'</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">DataLoader</span><span class="o">.</span><span class="n">from_h5ad</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_data_from_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"Unsupported file type. Select .pt or .txt or .h5ad"</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">shape</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">num_u</span> <span class="o">=</span> <span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">num_v</span> <span class="o">=</span> <span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">"Data loaded successfully. Shape: "</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_u</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_v</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="SCOTCH.add_data_from_adata">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.add_data_from_adata">[docs]</a>
    <span class="k">def</span> <span class="nf">add_data_from_adata</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Loads data from AnnData object into SCOTCH framework.</span>

<span class="sd">        :param adata: anndata.AnnData object to extract data from. Transforms adata.X to PyTorch object.</span>
<span class="sd">        :type adata: anndata.AnnData</span>
<span class="sd">        """</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"adata must be an AnnData object"</span><span class="p">)</span>

        <span class="c1"># Extract the X matrix and covert to a torch tensor</span>
        <span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
        <span class="k">if</span> <span class="n">issparse</span><span class="p">(</span><span class="n">X</span><span class="p">):</span>
            <span class="n">X_coo</span> <span class="o">=</span> <span class="n">X</span><span class="o">.</span><span class="n">tocoo</span><span class="p">()</span>
            <span class="n">values</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">X_coo</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
            <span class="n">indices</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">X_coo</span><span class="o">.</span><span class="n">row</span><span class="p">,</span> <span class="n">X_coo</span><span class="o">.</span><span class="n">col</span><span class="p">)))</span>
            <span class="n">X_tensor</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">sparse_coo_tensor</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">values</span><span class="p">,</span> <span class="n">X_coo</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
            <span class="n">X_tensor</span> <span class="o">=</span> <span class="n">X_tensor</span><span class="o">.</span><span class="n">to_dense</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">X_tensor</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">X_tensor</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_u</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_v</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">return</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="SCOTCH.add_scotch_embeddings_to_adata">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.add_scotch_embeddings_to_adata">[docs]</a>
    <span class="k">def</span> <span class="nf">add_scotch_embeddings_to_adata</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s2">""</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Adds SCOTCH objects to an AnnData object.</span>

<span class="sd">        :param prefix: Prefix to add to AnnData objects created by SCOTCH.</span>
<span class="sd">        :type prefix: str</span>

<span class="sd">        :param adata: The AnnData object to which SCOTCH embeddings will be added.</span>
<span class="sd">        :type adata: anndata.AnnData</span>
<span class="sd">        """</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"adata must be an AnnData object"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"prefix must be a string"</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">prefix</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">'_'</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s1">'_'</span>

        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s1">'cell_clusters'</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">U_assign</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">"gene_clusters"</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">V_assign</span><span class="o">.</span><span class="n">t</span><span class="p">()</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s1">'cell_embedding'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">U</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s1">'gene_embedding'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">V</span><span class="o">.</span><span class="n">t</span><span class="p">()</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s1">'S_matrix'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">S</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s1">'P_embedding'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">P</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s1">'Q_embedding'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">Q</span><span class="o">.</span><span class="n">t</span><span class="p">()</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s1">'reconstruction_error'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reconstruction_error</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s1">'error'</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">adata</span></div>


<div class="viewcode-block" id="SCOTCH.make_adata_from_scotch">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.make_adata_from_scotch">[docs]</a>
    <span class="k">def</span> <span class="nf">make_adata_from_scotch</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s2">""</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Create an AnnData object from the given data.</span>

<span class="sd">        :param self: The instance of the class containing the data.</span>
<span class="sd">        :type self: object</span>

<span class="sd">        :param prefix: A string appended to the generated AnnData objects.</span>
<span class="sd">        :type prefix: str</span>

<span class="sd">        :returns: An AnnData object containing the processed data.</span>
<span class="sd">        :rtype: anndata.AnnData</span>
<span class="sd">        """</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"prefix must be a str"</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">prefix</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">'_'</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s1">'_'</span>
        <span class="n">X</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">X</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">add_scotch_embeddings_to_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">prefix</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">adata</span></div>


<div class="viewcode-block" id="SCOTCH.make_top_regulators_list">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.make_top_regulators_list">[docs]</a>
    <span class="k">def</span> <span class="nf">make_top_regulators_list</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">gene_cluster_id</span><span class="o">=</span><span class="s2">"gene_clusters"</span><span class="p">,</span> <span class="n">gene_embedding_id</span><span class="o">=</span><span class="s2">"gene_embedding"</span><span class="p">,</span>
                                 <span class="n">prefix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">top_k</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">            Create a list of top regulators for each gene cluster based on gene embeddings.</span>

<span class="sd">            :param self: The instance of the class containing the data.</span>
<span class="sd">            :type self: object</span>

<span class="sd">            :param adata: An AnnData object containing single-cell gene expression data.</span>
<span class="sd">            :type adata: anndata.AnnData</span>

<span class="sd">            :param gene_cluster_id: The key for the gene clusters stored in `adata.var`. (default is "gene_clusters")</span>
<span class="sd">            :type gene_cluster_id: str</span>

<span class="sd">            :param gene_embedding_id: The key for the gene embedding matrix stored in `adata.varm`. (default is "gene_embedding")</span>
<span class="sd">            :type gene_embedding_id: str</span>

<span class="sd">            :param prefix: The string utilized when adding SCOTCH data to anndata. Use instead of gene_cluster_id and gene_embedding_id. (default is None)</span>
<span class="sd">            :type prefix: str</span>

<span class="sd">            :param top_k: The number of top genes to select per cluster (default is 5).</span>
<span class="sd">            :type top_k: int, optional</span>

<span class="sd">            :returns: A list of tuples, each containing the cluster index and the top `top_k` genes for that cluster.</span>
<span class="sd">            :rtype: list of tuples</span>
<span class="sd">        """</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="n">prefix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"prefix must be a string if passed."</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">prefix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">prefix</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">'_'</span><span class="p">:</span>
                <span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s1">'_'</span>
            <span class="n">gene_cluster_id</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">gene_cluster_id</span>
            <span class="n">gene_embedding_id</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">gene_embedding_id</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"adata must be an AnnData object"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_cluster_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"gene_cluster_id must be a string"</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">gene_cluster_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"gene_cluster_id '</span><span class="si">{</span><span class="n">gene_embedding_id</span><span class="si">}</span><span class="s2">' not found in adata.var.columns"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_embedding_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"gene_embedding_id must be a string"</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">gene_embedding_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"gene_embedding_id '</span><span class="si">{</span><span class="n">gene_embedding_id</span><span class="si">}</span><span class="s2">' not found in adata.varm.keys()"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">top_k</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"top_k must be an integer"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">top_k</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"top_k must greater than zero"</span><span class="p">)</span>

        <span class="n">cluster_gene</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">k2</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="n">gene_embedding_id</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">k2</span><span class="p">):</span>
            <span class="n">adata_slice</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">]</span> <span class="o">==</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">adata_slice</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">top_k</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">"Warning: only </span><span class="si">{</span><span class="n">adata_slice</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> genes are assigned to cluster </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">. Including only these genes"</span><span class="p">)</span>
                <span class="n">top_slice</span> <span class="o">=</span> <span class="n">adata_slice</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">top_slice</span> <span class="o">=</span> <span class="n">top_k</span>
            <span class="n">embedding</span> <span class="o">=</span> <span class="n">adata_slice</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="n">gene_embedding_id</span><span class="p">]</span>
            <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">embedding</span><span class="p">[:,</span> <span class="n">i</span><span class="p">])[</span><span class="o">-</span><span class="n">top_slice</span><span class="p">:,</span> <span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">top_genes</span> <span class="o">=</span> <span class="n">adata_slice</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">indices</span><span class="p">]</span>
            <span class="n">cluster_gene</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">i</span><span class="p">,</span> <span class="n">top_genes</span><span class="p">))</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">c_g</span> <span class="ow">in</span> <span class="n">cluster_gene</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Gene Cluster </span><span class="si">{</span><span class="n">c_g</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">c_g</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cluster_gene</span></div>


<div class="viewcode-block" id="SCOTCH.run_enrich_analyzer">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.run_enrich_analyzer">[docs]</a>
    <span class="k">def</span> <span class="nf">run_enrich_analyzer</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">gene_cluster_id</span><span class="p">,</span> <span class="n">go_regnet_file</span><span class="p">,</span> <span class="n">fdr</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">test_type</span><span class="o">=</span><span class="s1">'persg'</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s1">'GO'</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">            Perform gene ontology (GO) enrichment analysis and store results in the AnnData object.</span>

<span class="sd">            :param self: The instance of the class containing the data.</span>
<span class="sd">            :type self: object</span>

<span class="sd">            :param adata: An AnnData object containing single-cell gene expression data.</span>
<span class="sd">            :type adata: anndata.AnnData</span>

<span class="sd">            :param gene_cluster_id: Identifier for the gene cluster to be analyzed.</span>
<span class="sd">            :type gene_cluster_id: str</span>

<span class="sd">            :param go_regnet_file: File path to the gene ontology (GO) enrichment file.</span>
<span class="sd">            :type go_regnet_file: str</span>

<span class="sd">            :param fdr: The false discovery rate threshold (default is 0.05).</span>
<span class="sd">            :type fdr: float, optional</span>

<span class="sd">            :param test_type: Type of statistical test to be performed (default is 'persg'). Valid options are 'persg' and 'fullgraph'.</span>
<span class="sd">            :type test_type: str, optional</span>

<span class="sd">            :param prefix: Prefix for storing enrichment results in the AnnData object (default is 'GO').</span>
<span class="sd">            :type prefix: str, optional</span>

<span class="sd">            :returns: None. Enrichment results are stored in `adata.uns[prefix + "enrichment"]`.</span>
<span class="sd">            :rtype: None</span>
<span class="sd">        """</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"adata must be an AnnData object"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_cluster_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"gene_cluster_id must be a string"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">gene_cluster_id</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"gene_cluster_id '</span><span class="si">{</span><span class="n">gene_cluster_id</span><span class="si">}</span><span class="s2">' not found in adata.var"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">go_regnet_file</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"go_regnet_file must be a string"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">go_regnet_file</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The go_regnet_file '</span><span class="si">{</span><span class="n">go_regnet_file</span><span class="si">}</span><span class="s2">' does not exist"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">go_regnet_file</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">R_OK</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The go_regnet_file '</span><span class="si">{</span><span class="n">go_regnet_file</span><span class="si">}</span><span class="s2">' is not readable"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">fdr</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"fdr must be a float"</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">fdr</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">fdr</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"fdr must between 0 and 1"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">test_type</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"test_type must be a string"</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">test_type</span> <span class="o">!=</span> <span class="s2">"persg"</span> <span class="ow">and</span> <span class="n">test_type</span> <span class="o">!=</span> <span class="s2">"fullgraph"</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"test_type must be 'persg' or 'fullgraph'"</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"prefix must be a string"</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">prefix</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">'_'</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s1">'_'</span>


        <span class="n">EA</span> <span class="o">=</span> <span class="n">pyEnrichAnalyzer</span><span class="o">.</span><span class="n">Framework</span><span class="p">()</span>
        <span class="n">enrichment</span> <span class="o">=</span> <span class="n">EA</span><span class="o">.</span><span class="n">runEnrichAnalyzer</span><span class="p">(</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="n">orient</span><span class="o">=</span><span class="s1">'index'</span><span class="p">),</span>
            <span class="n">gene_cluster_id</span><span class="p">,</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span>
            <span class="n">go_regnet_file</span><span class="p">,</span>
            <span class="n">fdr</span><span class="p">,</span>
            <span class="n">test_type</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">enrichment</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s1">'SubGraphName'</span><span class="p">:</span> <span class="s1">'gene cluster'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s1">'enrichment'</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span></div>


<div class="viewcode-block" id="SCOTCH.visualize_enrichment_bubbleplots">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.visualize_enrichment_bubbleplots">[docs]</a>
    <span class="k">def</span> <span class="nf">visualize_enrichment_bubbleplots</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">enrich_object_id</span><span class="p">,</span>
                                         <span class="n">gene_cluster_id</span><span class="o">=</span><span class="s1">'gene cluster'</span><span class="p">,</span>
                                         <span class="n">term_id</span><span class="o">=</span><span class="s1">'TermName'</span><span class="p">,</span>
                                         <span class="n">FC_id</span><span class="o">=</span><span class="s1">'Foldenr'</span><span class="p">,</span>
                                         <span class="n">q_val_id</span><span class="o">=</span><span class="s1">'CorrPval'</span><span class="p">,</span>
                                         <span class="n">top_k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                                         <span class="n">max_point_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                                         <span class="n">palette</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span>
                                         <span class="n">gene_cluster_set</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                         <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Visualize enrichment results as a bubble plot, where the size of the bubbles represents log2 fold change</span>
<span class="sd">        and the color represents -log10 p-values.</span>

<span class="sd">        :param self: The instance of the class containing the data.</span>
<span class="sd">        :type self: object</span>

<span class="sd">        :param adata: An AnnData object containing single-cell gene expression data.</span>
<span class="sd">        :type adata: anndata.AnnData</span>

<span class="sd">        :param enrich_object_id: The key in `adata.uns` containing the enrichment data.</span>
<span class="sd">        :type enrich_object_id: str</span>

<span class="sd">        :param gene_cluster_id: The column name representing gene clusters in the enrichment data (default is 'gene cluster').</span>
<span class="sd">        :type gene_cluster_id: str, optional</span>

<span class="sd">        :param term_id: The column name representing terms in the enrichment data (default is 'TermName').</span>
<span class="sd">        :type term_id: str, optional</span>

<span class="sd">        :param FC_id: The column name representing fold change values in the enrichment data (default is 'Foldenr').</span>
<span class="sd">        :type FC_id: str, optional</span>

<span class="sd">        :param q_val_id: The column name representing the corrected p-values in the enrichment data (default is 'CorrPval').</span>
<span class="sd">        :type q_val_id: str, optional</span>

<span class="sd">        :param top_k: The number of top terms to select per gene cluster (default is 5).</span>
<span class="sd">        :type top_k: int, optional</span>

<span class="sd">        :param max_point_size: The maximum size of the bubbles in the plot (default is 100).</span>
<span class="sd">        :type max_point_size: int, optional</span>

<span class="sd">        :param palette: The color palette used for the plot (default is 'viridis').</span>
<span class="sd">        :type palette: str, optional</span>

<span class="sd">        :returns: A `matplotlib.figure.Figure` object containing the bubble plot.</span>
<span class="sd">        :rtype: matplotlib.figure.Figure</span>
<span class="sd">        """</span>

        <span class="n">enrichment</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">enrich_object_id</span><span class="p">]</span>
        <span class="n">enrichment_topk_terms</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">gene_cluster_id</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">grp</span><span class="p">:</span> <span class="n">grp</span><span class="o">.</span><span class="n">nsmallest</span><span class="p">(</span><span class="n">top_k</span><span class="p">,</span> <span class="n">q_val_id</span><span class="p">),</span>
                                                                          <span class="n">include_groups</span><span class="o">=</span><span class="kc">False</span><span class="p">)[</span><span class="n">term_id</span><span class="p">]</span>
        <span class="n">filtered_enrichment</span> <span class="o">=</span> <span class="n">enrichment</span><span class="p">[</span><span class="n">enrichment</span><span class="p">[</span><span class="n">term_id</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">enrichment_topk_terms</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">filtered_enrichment</span><span class="p">[</span><span class="s2">"log2FC"</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">filtered_enrichment</span><span class="p">[</span><span class="n">FC_id</span><span class="p">])</span>
        <span class="n">filtered_enrichment</span><span class="p">[</span><span class="s2">"-log10QVal"</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">filtered_enrichment</span><span class="p">[</span><span class="n">q_val_id</span><span class="p">])</span>

        <span class="n">filtered_enrichment</span><span class="p">[</span><span class="n">term_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">filtered_enrichment</span><span class="p">[</span><span class="n">term_id</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">gene_cluster_set</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">filtered_enrichment</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">filtered_enrichment</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">],</span>
                                                                  <span class="n">categories</span><span class="o">=</span><span class="n">gene_cluster_set</span><span class="p">,</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">filtered_enrichment</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">filtered_enrichment</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">])</span>

        <span class="n">filtered_enrichment</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fill_missing_categories</span><span class="p">(</span><span class="n">filtered_enrichment</span><span class="p">,</span>
                                                            <span class="n">categorical_columns</span><span class="o">=</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">,</span> <span class="n">term_id</span><span class="p">],</span>
                                                            <span class="n">fill_value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">categories</span> <span class="o">=</span> <span class="n">filtered_enrichment</span><span class="p">[</span><span class="n">gene_cluster_id</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>

        <span class="n">fig</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">))</span>

        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span>
            <span class="n">data</span><span class="o">=</span><span class="n">filtered_enrichment</span><span class="p">,</span>
            <span class="n">x</span><span class="o">=</span><span class="n">gene_cluster_id</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="n">term_id</span><span class="p">,</span>
            <span class="n">size</span><span class="o">=</span><span class="s1">'log2FC'</span><span class="p">,</span>
            <span class="n">hue</span><span class="o">=</span><span class="s1">'-log10QVal'</span><span class="p">,</span>
            <span class="n">sizes</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_point_size</span><span class="p">),</span>
            <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span>
            <span class="n">zorder</span><span class="o">=</span><span class="mi">3</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">categories</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">fig</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Gene Cluster'</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">fig</span></div>


<div class="viewcode-block" id="SCOTCH.visualize_marker_gene_bubbleplot_per_cell_cluster">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.visualize_marker_gene_bubbleplot_per_cell_cluster">[docs]</a>
    <span class="k">def</span> <span class="nf">visualize_marker_gene_bubbleplot_per_cell_cluster</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span>
                                                          <span class="n">cell_cluster_id</span><span class="o">=</span><span class="s1">'cell_clusters'</span><span class="p">,</span>
                                                          <span class="n">gene_cluster_id</span><span class="o">=</span><span class="s1">'gene_clusters'</span><span class="p">,</span>
                                                          <span class="n">gene_embedding_id</span><span class="o">=</span><span class="s1">'gene_embedding'</span><span class="p">,</span>
                                                          <span class="n">prefix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                                          <span class="n">top_k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                                                          <span class="n">max_point_size</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
                                                          <span class="n">palette</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span>
                                                          <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Visualize marker gene expression as a bubble plot, where the size of the bubbles represents the</span>
<span class="sd">        percent of non-zero counts and the color represents the mean marker expression. Top_k are selected for each</span>
<span class="sd">        gene cluster. Genes names are follows by the gene cluster that each gene corresponds to.</span>

<span class="sd">        :param self: The instance of the class containing the data.</span>
<span class="sd">        :type self: object</span>

<span class="sd">        :param adata: An AnnData object containing single-cell gene expression data.</span>
<span class="sd">        :type adata: anndata.AnnData</span>

<span class="sd">        :param cell_cluster_id: The column name representing cell clusters in `adata.obs`. (default is 'cell_clusters')</span>
<span class="sd">        :type cell_cluster_id: str</span>

<span class="sd">        :param gene_cluster_id: The column name representing gene clusters in `adata.var`. (default is 'gene_clusters')</span>
<span class="sd">        :type gene_cluster_id: str</span>

<span class="sd">        :param gene_embedding_id: The identifier for the gene embedding used for selecting top markers. (default is 'gene_embedding')</span>
<span class="sd">        :type gene_embedding_id: str</span>

<span class="sd">        :param prefix: The string utilized when adding SCOTCH data to anndata. Use instead of gene_cluster_id and gene_embedding_id. (default is None)</span>
<span class="sd">        :type prefix: str</span>

<span class="sd">        :param top_k: The number of top markers to consider per gene cluster (default is 5).</span>
<span class="sd">        :type top_k: int, optional</span>

<span class="sd">        :param max_point_size: The maximum size of the bubbles in the plot (default is 300).</span>
<span class="sd">        :type max_point_size: int, optional</span>

<span class="sd">        :param palette: The color palette used for the plot (default is 'viridis').</span>
<span class="sd">        :type palette: str, optional</span>

<span class="sd">        :param ax: The matplotlib.axes.axes object to plot in. If none, new figure is generated and returned (default is None)</span>
<span class="sd">        :type ax: matplotlib.axes.Axes, optional</span>

<span class="sd">        :returns: A `matplotlib.figure.Figure` object containing the bubble plot if ax not passed. Else returns none.</span>
<span class="sd">        :rtype: matplotlib.figure.Figure or None</span>
<span class="sd">        """</span>
        <span class="k">if</span> <span class="n">prefix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`prefix` must be a string if provided."</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">prefix</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">'_'</span><span class="p">:</span>
                <span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s1">'_'</span>
            <span class="n">cell_cluster_id</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">cell_cluster_id</span>
            <span class="n">gene_cluster_id</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">gene_cluster_id</span>
            <span class="n">gene_embedding_id</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">gene_embedding_id</span>

        <span class="c1"># Safety checks for `adata`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`adata` must be an AnnData object."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `cell_cluster_id`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cell_cluster_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`cell_cluster_id` must be a string."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">cell_cluster_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"`cell_cluster_id` '</span><span class="si">{</span><span class="n">cell_cluster_id</span><span class="si">}</span><span class="s2">' was not found in `adata.obs.columns`."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `gene_cluster_id`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_cluster_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`gene_cluster_id` must be a string."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">gene_cluster_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"`gene_cluster_id` '</span><span class="si">{</span><span class="n">gene_cluster_id</span><span class="si">}</span><span class="s2">' was not found in `adata.var.columns`."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `gene_embedding_id`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_embedding_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`gene_embedding_id` must be a string."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">gene_embedding_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"`gene_embedding_id` '</span><span class="si">{</span><span class="n">gene_embedding_id</span><span class="si">}</span><span class="s2">' was not found in `adata.varm.keys()`."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `top_k`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">top_k</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`top_k` must be an integer."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">top_k</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"`top_k` must be greater than zero."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `max_point_size`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">max_point_size</span><span class="p">,</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`max_point_size` must be a number (int or float)."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">max_point_size</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"`max_point_size` must be greater than zero."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `palette`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">palette</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`palette` must be a string."</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">palette</span><span class="p">)</span>  <span class="c1"># Check if palette exists</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"'</span><span class="si">{</span><span class="n">palette</span><span class="si">}</span><span class="s2">' is not a valid color palette. Please choose a valid matplotlib colormap."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `ax`</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">plt</span><span class="o">.</span><span class="n">Axes</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`ax` must be a matplotlib.axes.Axes object or None."</span><span class="p">)</span>

        <span class="n">markers</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_top_regulators_list</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span> <span class="n">gene_cluster_id</span><span class="o">=</span><span class="n">gene_cluster_id</span><span class="p">,</span>
                                                <span class="n">gene_embedding_id</span><span class="o">=</span><span class="n">gene_embedding_id</span><span class="p">,</span> <span class="n">top_k</span><span class="o">=</span><span class="n">top_k</span><span class="p">)</span>

        <span class="n">marker_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">top_gene</span> <span class="ow">in</span> <span class="n">markers</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">top_gene</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="n">marker_list_idx</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">marker_list</span><span class="p">)</span>

        <span class="n">marker_expression</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">marker_list_idx</span><span class="p">]</span>

        <span class="n">mean_marker_expression</span> <span class="o">=</span> <span class="n">marker_expression</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_cluster_id</span><span class="p">],</span> <span class="n">observed</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">percent_marker_expression</span> <span class="o">=</span> <span class="p">(</span><span class="n">marker_expression</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_cluster_id</span><span class="p">],</span>
                                                                            <span class="n">observed</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>

        <span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">marker</span> <span class="ow">in</span> <span class="n">markers</span><span class="p">:</span>
            <span class="n">gene_cluster</span> <span class="o">=</span> <span class="n">marker</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="n">marker</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

            <span class="c1"># Iterate over the genes</span>
            <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
                <span class="c1"># Get the mean expression for this gene in this cluster</span>
                <span class="n">mean_exp</span> <span class="o">=</span> <span class="n">mean_marker_expression</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">gene</span><span class="p">]</span>

                <span class="c1"># Get the percent non-zero expression for this gene in this cluster</span>
                <span class="n">non_zero</span> <span class="o">=</span> <span class="n">percent_marker_expression</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">gene</span><span class="p">]</span>

                <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mean_marker_expression</span><span class="o">.</span><span class="n">index</span><span class="p">):</span>
                    <span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">gene_cluster</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">mean_exp</span><span class="p">[</span><span class="n">idx</span><span class="p">],</span> <span class="n">cluster</span><span class="p">,</span> <span class="n">non_zero</span><span class="p">[</span><span class="n">idx</span><span class="p">]])</span>

        <span class="c1"># Convert the data list into a DataFrame</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">"Gene Cluster"</span><span class="p">,</span> <span class="s2">"Gene"</span><span class="p">,</span> <span class="s2">"Mean Marker Expression"</span><span class="p">,</span> <span class="s2">"Cell Cluster"</span><span class="p">,</span>
                                         <span class="s2">"Percent Nonzero Count"</span><span class="p">])</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">'Gene Cluster'</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">'Gene Cluster'</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">'Cell Cluster'</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">'Cell Cluster'</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'string'</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">'Gene'</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">'Gene'</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">'category'</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">"Gene Cluster"</span><span class="p">,</span> <span class="s1">'Gene'</span><span class="p">])</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">"Gene_Gene_Cluster"</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="sa">f</span><span class="s1">'</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">"Gene"</span><span class="p">]</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s2">"Gene Cluster"</span><span class="p">]</span><span class="si">}</span><span class="s1">'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="n">max_expression</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">'Mean Marker Expression'</span><span class="p">]))</span>

        <span class="n">fig</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">))</span>

        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span>
            <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
            <span class="n">x</span><span class="o">=</span><span class="s1">'Cell Cluster'</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="s1">'Gene_Gene_Cluster'</span><span class="p">,</span>
            <span class="n">hue</span><span class="o">=</span><span class="s1">'Mean Marker Expression'</span><span class="p">,</span>
            <span class="n">size</span><span class="o">=</span><span class="s1">'Percent Nonzero Count'</span><span class="p">,</span>
            <span class="n">sizes</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_point_size</span><span class="p">),</span>  <span class="c1"># control size range of bubbles</span>
            <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span>  <span class="c1"># color palette</span>
            <span class="n">edgecolor</span><span class="o">=</span><span class="s1">'w'</span><span class="p">,</span>  <span class="c1"># white edge for visibility</span>
            <span class="n">hue_norm</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_expression</span><span class="p">),</span>
            <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
            <span class="n">zorder</span><span class="o">=</span><span class="mi">3</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">fig</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Adding colorbar and labels</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">'Cell Cluster'</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">fig</span></div>


<div class="viewcode-block" id="SCOTCH.visualize_marker_gene_bubbleplot_per_gene_cluster">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.visualize_marker_gene_bubbleplot_per_gene_cluster">[docs]</a>
    <span class="k">def</span> <span class="nf">visualize_marker_gene_bubbleplot_per_gene_cluster</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span>
                                                          <span class="n">gene_cluster_id</span><span class="o">=</span><span class="s1">'gene_clusters'</span><span class="p">,</span>
                                                          <span class="n">gene_embedding_id</span><span class="o">=</span><span class="s2">"gene_embedding"</span><span class="p">,</span>
                                                          <span class="n">prefix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                                          <span class="n">top_k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                                                          <span class="n">max_point_size</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
                                                          <span class="n">palette</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span>
                                                          <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Visualize marker gene expression as a bubble plot, where the size of the bubbles represents the</span>
<span class="sd">        percent of non-zero counts and the color represents the mean marker expression. Top_k are selected for each</span>
<span class="sd">        gene cluster. Genes names are follows by the gene cluster that each gene corresponds to.</span>

<span class="sd">        :param self: The instance of the class containing the data.</span>
<span class="sd">        :type self: object</span>

<span class="sd">        :param adata: An AnnData object containing single-cell gene expression data.</span>
<span class="sd">        :type adata: anndata.AnnData</span>

<span class="sd">        :param gene_cluster_id: The identifier for the gene cluster used for selecting top markers.</span>
<span class="sd">        :type gene_cluster_id: str, optional</span>

<span class="sd">        :param gene_embedding_id: The identifier for the gene embedding used for selecting top markers.</span>
<span class="sd">        :type gene_embedding_id: str, optional</span>

<span class="sd">        :param prefix: The prefix string used when adding SCOTCH data to anndata object.</span>
<span class="sd">        :type prefix: str, optional</span>

<span class="sd">        :param top_k: The number of top markers to consider per gene cluster (default is 5).</span>
<span class="sd">        :type top_k: int, optional</span>

<span class="sd">        :param max_point_size: The maximum size of the bubbles in the plot (default is 300).</span>
<span class="sd">        :type max_point_size: int, optional</span>

<span class="sd">        :param palette: The color palette used for the plot (default is 'viridis').</span>
<span class="sd">        :type palette: str, optional</span>

<span class="sd">        :param ax: A `matplotlib.axes.Axes` object to plot the bubble plot. If not provided a new figures is generated</span>
<span class="sd">        :type ax: matplotlib.axes.Axes, optional</span>

<span class="sd">        :returns: A `matplotlib.figure.Figure` object containing the bubble plot.</span>
<span class="sd">        :rtype: matplotlib.figure.Figure</span>
<span class="sd">        """</span>

        <span class="k">if</span> <span class="n">prefix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">prefix</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`prefix` must be a string if provided."</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">prefix</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">'_'</span><span class="p">:</span>
                <span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s1">'_'</span>
            <span class="n">gene_cluster_id</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">gene_cluster_id</span>
            <span class="n">gene_embedding_id</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">gene_embedding_id</span>

        <span class="c1"># Safety checks for `adata`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`adata` must be an AnnData object."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `gene_cluster_id`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_cluster_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`gene_cluster_id` must be a string."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">gene_cluster_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"`gene_cluster_id` '</span><span class="si">{</span><span class="n">gene_cluster_id</span><span class="si">}</span><span class="s2">' was not found in `adata.var.columns`."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `gene_embedding_id`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_embedding_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`gene_embedding_id` must be a string."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">gene_embedding_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">varm</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"`gene_embedding_id` '</span><span class="si">{</span><span class="n">gene_embedding_id</span><span class="si">}</span><span class="s2">' was not found in `adata.varm` keys."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `top_k`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">top_k</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`top_k` must be an integer."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">top_k</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"`top_k` must be greater than zero."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `max_point_size`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">max_point_size</span><span class="p">,</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`max_point_size` must be a number (int or float)."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">max_point_size</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"`max_point_size` must be greater than zero."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `palette`</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">palette</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`palette` must be a string."</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Check if palette is a valid matplotlib colormap</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">palette</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"'</span><span class="si">{</span><span class="n">palette</span><span class="si">}</span><span class="s2">' is not a valid color palette. Please choose a valid matplotlib colormap."</span><span class="p">)</span>

        <span class="c1"># Safety checks for `ax`</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">plt</span><span class="o">.</span><span class="n">Axes</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`ax` must be a matplotlib.axes.Axes object or None."</span><span class="p">)</span>

        <span class="n">markers</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_top_regulators_list</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span> <span class="n">gene_cluster_id</span><span class="o">=</span><span class="n">gene_cluster_id</span><span class="p">,</span>
                                                <span class="n">gene_embedding_id</span><span class="o">=</span><span class="n">gene_embedding_id</span><span class="p">,</span> <span class="n">top_k</span><span class="o">=</span><span class="n">top_k</span><span class="p">)</span>

        <span class="n">marker_list</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="n">gene</span> <span class="k">for</span> <span class="n">top_gene</span> <span class="ow">in</span> <span class="n">markers</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">top_gene</span><span class="p">[</span><span class="mi">1</span><span class="p">]])</span>
        <span class="n">marker_list_idx</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">marker_list</span><span class="p">)</span>

        <span class="n">marker_data</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">marker_list_idx</span><span class="p">]</span>
        <span class="n">marker_V_matrix</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">marker_data</span><span class="o">.</span><span class="n">varm</span><span class="p">[</span><span class="n">gene_embedding_id</span><span class="p">])</span>
        <span class="n">marker_V_matrix</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">marker_data</span><span class="o">.</span><span class="n">var_names</span>
        <span class="n">marker_V_matrix</span> <span class="o">=</span> <span class="n">marker_V_matrix</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
        <span class="n">marker_V_matrix</span> <span class="o">=</span> <span class="n">marker_V_matrix</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">'index'</span><span class="p">],</span> <span class="n">var_name</span><span class="o">=</span><span class="s2">"gene cluster"</span><span class="p">,</span> <span class="n">value_name</span><span class="o">=</span><span class="s1">'embedding'</span><span class="p">)</span>
        <span class="n">marker_V_order</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">marker_V_matrix</span><span class="p">[</span><span class="s1">'gene cluster'</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()]</span>
        <span class="n">marker_V_matrix</span><span class="p">[</span><span class="s1">'gene cluster'</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">marker_V_matrix</span><span class="p">[</span><span class="s1">'gene cluster'</span><span class="p">]],</span>
                                                         <span class="n">categories</span><span class="o">=</span><span class="n">marker_V_order</span><span class="p">,</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">marker_V_matrix</span><span class="p">[</span><span class="s1">'index'</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">marker_V_matrix</span><span class="p">[</span><span class="s1">'index'</span><span class="p">],</span> <span class="n">categories</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">marker_list</span><span class="p">),</span>
                                                  <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">max_embedding</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">marker_V_matrix</span><span class="p">[</span><span class="s1">'embedding'</span><span class="p">]))</span>

        <span class="n">categories</span> <span class="o">=</span> <span class="n">marker_V_matrix</span><span class="p">[</span><span class="s1">'gene cluster'</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>

        <span class="n">fig</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">))</span>

        <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span>
            <span class="n">data</span><span class="o">=</span><span class="n">marker_V_matrix</span><span class="p">,</span>
            <span class="n">x</span><span class="o">=</span><span class="s1">'gene cluster'</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="s1">'index'</span><span class="p">,</span>
            <span class="n">hue</span><span class="o">=</span><span class="s1">'embedding'</span><span class="p">,</span>
            <span class="n">size</span><span class="o">=</span><span class="s1">'embedding'</span><span class="p">,</span>
            <span class="n">sizes</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_point_size</span><span class="p">),</span>  <span class="c1"># control size range of bubbles</span>
            <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span>  <span class="c1"># color palette</span>
            <span class="n">edgecolor</span><span class="o">=</span><span class="s1">'w'</span><span class="p">,</span>  <span class="c1"># white edge for visibility</span>
            <span class="n">hue_norm</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_embedding</span><span class="p">),</span>
            <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span>
            <span class="n">zorder</span><span class="o">=</span><span class="mi">3</span>
        <span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">categories</span><span class="p">)</span> <span class="o">-</span> <span class="mf">0.5</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">fig</span></div>


<div class="viewcode-block" id="SCOTCH.plot_cooccurrence_proportions">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.plot_cooccurrence_proportions">[docs]</a>
    <span class="k">def</span> <span class="nf">plot_cooccurrence_proportions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">field_1</span><span class="o">=</span><span class="s2">"cell_clusters"</span><span class="p">,</span> <span class="n">field_2</span><span class="o">=</span><span class="s2">"sample"</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">"Reds"</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Generate a heatmap of co-occurrence proportions between two categorical variables.</span>

<span class="sd">        This function creates a heatmap to visualize the co-occurrence proportions of two categorical variables</span>
<span class="sd">        stored in the `adata.obs` dataframe. The rows correspond to values from `field_1` and the</span>
<span class="sd">        columns correspond to values from `field_2`. The heatmap displays normalized proportions per row.</span>

<span class="sd">        :param adata: An AnnData object containing the single-cell data.</span>
<span class="sd">            It must include `adata.obs` with `field_1` and `field_2` as categorical variables.</span>
<span class="sd">        :type adata: anndata.AnnData</span>

<span class="sd">        :param field_1: The name of the first categorical variable in `adata.obs` (used for heatmap rows).</span>
<span class="sd">            Its values will define the rows in the heatmap. (default is 'cell_clusters')</span>
<span class="sd">        :type field_1: str, optional</span>

<span class="sd">        :param field_2: The name of the second categorical variable in `adata.obs` (used for heatmap columns).</span>
<span class="sd">            Its values will define the columns in the heatmap. (Default is 'sample'.)</span>
<span class="sd">        :type field_2: str, optional</span>

<span class="sd">        :param cmap: The color map used for the heatmap. It must be a valid Matplotlib colormap, with "Reds"</span>
<span class="sd">            as the default. This determines the gradient colors representing value intensity in the heatmap.</span>
<span class="sd">        :type cmap: str, optional</span>

<span class="sd">        :param ax: A Matplotlib `Axes` object on which the heatmap will be plotted. If `None`, a new figure</span>
<span class="sd">            and axis are created, and the function returns the generated figure. If provided, the heatmap</span>
<span class="sd">            is plotted on the existing axis, and no figure is returned.</span>
<span class="sd">        :type ax: matplotlib.axes.Axes or None, optional</span>

<span class="sd">        :returns: A Matplotlib `Figure` object containing the heatmap of co-occurrence proportions,</span>
<span class="sd">            if a new figure is generated. If `ax` is provided, the function returns None.</span>
<span class="sd">        :rtype: matplotlib.figure.Figure or None</span>

<span class="sd">        :raises TypeError: If `adata` is not an `AnnData` object, or `field_1` and `field_2` are not strings.</span>
<span class="sd">        :raises ValueError: If `field_1` or `field_2` are not found in `adata.obs.columns`.</span>
<span class="sd">        """</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s1">'adata must be an Adata object'</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">field_1</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s1">'variable_1 must be a string'</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">field_2</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s1">'variable_2 must be a string'</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">field_1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"'</span><span class="si">{</span><span class="n">field_1</span><span class="si">}</span><span class="s2">' not found in adata.obs."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">field_2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"'</span><span class="si">{</span><span class="n">field_2</span><span class="si">}</span><span class="s2">' not found in adata.obs."</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cmap</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`cmap` must be a string representing a valid Matplotlib colormap."</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Check if the colormap is valid</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">cmap</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"'</span><span class="si">{</span><span class="n">cmap</span><span class="si">}</span><span class="s2">' is not a valid Matplotlib colormap."</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">plt</span><span class="o">.</span><span class="n">Axes</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`ax` must be a `matplotlib.axes.Axes` object or `None`."</span><span class="p">)</span>

        <span class="c1"># Extract the categorical variables from adata.obs</span>
        <span class="n">var_1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">field_1</span><span class="p">]</span>
        <span class="n">var_2</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">field_2</span><span class="p">]</span>

        <span class="c1"># Create a contingency table of co-occurrence counts</span>
        <span class="n">co_occurrence_counts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">crosstab</span><span class="p">(</span><span class="n">var_1</span><span class="p">,</span> <span class="n">var_2</span><span class="p">)</span>

        <span class="c1"># Normalize by rows to get proportions</span>
        <span class="n">co_occurrence_proportions</span> <span class="o">=</span> <span class="n">co_occurrence_counts</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">co_occurrence_counts</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># Plot the heatmap</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">))</span>

        <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">co_occurrence_proportions</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s1">'.2f'</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">fig</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">fig</span></div>


<div class="viewcode-block" id="SCOTCH.plot_element_count_heatmap">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.plot_element_count_heatmap">[docs]</a>
    <span class="k">def</span> <span class="nf">plot_element_count_heatmap</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">field</span><span class="o">=</span><span class="s1">'cell_clusters'</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s1">'vertical'</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">'Blues'</span><span class="p">,</span> <span class="n">v_min</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                                   <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        This function produces a heatmap displaying the count of unique elements in a specific column of an `AnnData` object.</span>
<span class="sd">        The orientation of the heatmap can be controlled, along with customization options like the color map and axis.</span>

<span class="sd">        :param adata: An AnnData object containing the single-cell data.</span>
<span class="sd">            This parameter stores observations and variables, including metadata used for the analysis.</span>
<span class="sd">        :type adata: anndata.AnnData</span>

<span class="sd">        :param field: The column in `adata.obs` for which the counts should be calculated.</span>
<span class="sd">            The unique values in this column are counted and visualized in the heatmap. Default is cell_clusters</span>
<span class="sd">        :type field: str optional.</span>

<span class="sd">        :param orientation: The orientation of the heatmap (either rows or columns represent the elements being counted).</span>
<span class="sd">            Acceptable values are 'vertical' (default) or 'horizontal'.</span>
<span class="sd">        :type orientation: str, optional</span>

<span class="sd">        :param cmap: The color map used to style the heatmap. For example, use 'Blues' for a blue shade gradient.</span>
<span class="sd">            Defaults to 'Blues'.</span>
<span class="sd">        :type cmap: str, optional</span>

<span class="sd">        :param v_min: The minimum value for the heatmap color scale.</span>
<span class="sd">            This is useful to set a threshold for visualization. Default is 0.</span>
<span class="sd">        :type v_min: int, optional</span>

<span class="sd">        :param ax: A matplotlib `Axes` object onto which the heatmap will be drawn.</span>
<span class="sd">            If not provided, a new figure and axes will be created.</span>
<span class="sd">        :type ax: matplotlib.axes.Axes, optional</span>

<span class="sd">        :return: A matplotlib `Figure` object containing the heatmap.</span>
<span class="sd">            If `ax` is provided, then the returned `Figure` will be `None`, as the plot will be drawn on the given `Axes`.</span>
<span class="sd">        :rtype: matplotlib.figure.Figure or None</span>
<span class="sd">        """</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`adata` must be an instance of `anndata.AnnData`."</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">field</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`field` must be a string."</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">field</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The field `</span><span class="si">{</span><span class="n">field</span><span class="si">}</span><span class="s2">` is not found in `adata.obs` columns."</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">"vertical"</span><span class="p">,</span> <span class="s2">"horizontal"</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"`orientation` must be either 'vertical' or 'horizontal'."</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">cmap</span><span class="p">)</span>  <span class="c1"># Access colormap to verify it's valid</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The colormap `</span><span class="si">{</span><span class="n">cmap</span><span class="si">}</span><span class="s2">` is not valid. Check available Matplotlib colormaps."</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">v_min</span><span class="p">,</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">))</span> <span class="ow">and</span> <span class="n">v_min</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"`v_min` must be a non-negative number (int or float)."</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">plt</span><span class="o">.</span><span class="n">Axes</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"`ax` must be an instance of matplotlib `Axes` or None. Received type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span><span class="si">}</span><span class="s2">."</span><span class="p">)</span>

        <span class="c1"># Count the occurrences of each unique element in the field</span>
        <span class="n">element_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">field</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="c1"># min_counts = min(element_counts)</span>
        <span class="n">max_counts</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">element_counts</span><span class="p">)</span>

        <span class="n">count_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">element_counts</span><span class="p">)</span>

        <span class="c1"># Create a heatmap based on the element counts</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">15</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">'vertical'</span><span class="p">:</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">count_df</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s1">'d'</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="n">v_min</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="n">max_counts</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">field</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">orientation</span> <span class="o">==</span> <span class="s1">'horizontal'</span><span class="p">:</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">count_df</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s1">'d'</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="n">v_min</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="n">max_counts</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">field</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"Orientation must be either 'vertical' or 'horizontal'."</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">fig</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Show the plot</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">fig</span></div>



<div class="viewcode-block" id="SCOTCH.plot_S_matrix">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.plot_S_matrix">[docs]</a>
    <span class="k">def</span> <span class="nf">plot_S_matrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">S_matrix_id</span><span class="o">=</span><span class="s2">"S_matrix"</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Plot a heatmap of the S matrix stored in `adata.uns` under the given key (`S_matrix_id`).</span>

<span class="sd">        :param adata: An AnnData object containing the single-cell data.</span>
<span class="sd">        :type adata: anndata.AnnData</span>

<span class="sd">        :param S_matrix_id: The key in `adata.uns` where the S matrix is stored. Default is 'S_matrix'.</span>
<span class="sd">        :type S_matrix_id: str</span>

<span class="sd">        :param palette: The color palette to use for the heatmap. Default is 'viridis'.</span>
<span class="sd">        :type palette: str</span>

<span class="sd">        :param ax: A matplotlib Axes object to plot the heatmap on. If not provided, a new figure and Axes will be created.</span>
<span class="sd">        :type ax: matplotlib.axes.Axes, optional</span>

<span class="sd">        :return: A matplotlib Figure object containing the heatmap, or None if `ax` is provided.</span>
<span class="sd">        :rtype: matplotlib.figure.Figure or None</span>
<span class="sd">        """</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`adata` must be an instance of `anndata.AnnData`."</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">S_matrix_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">"`S_matrix_id` must be a string."</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">S_matrix_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"`S_matrix_id` '</span><span class="si">{</span><span class="n">S_matrix_id</span><span class="si">}</span><span class="s2">' not found in `adata.uns`."</span><span class="p">)</span>

        <span class="n">S</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">S_matrix_id</span><span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">S</span><span class="p">,</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">))</span> <span class="ow">and</span> <span class="n">S</span><span class="o">.</span><span class="n">ndim</span> <span class="o">==</span> <span class="mi">2</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The object linked to `S_matrix_id` '</span><span class="si">{</span><span class="n">S_matrix_id</span><span class="si">}</span><span class="s2">' must be a 2D array or DataFrame."</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">palette</span><span class="p">)</span>  <span class="c1"># Will raise a ValueError if the colormap is invalid</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The colormap '</span><span class="si">{</span><span class="n">palette</span><span class="si">}</span><span class="s2">' is not a valid matplotlib colormap."</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">plt</span><span class="o">.</span><span class="n">Axes</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"If provided, `ax` must be a matplotlib `Axes` object. Received: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span><span class="si">}</span><span class="s2">."</span><span class="p">)</span>

        <span class="n">fig</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fix</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">S_matrix_id</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">:</span>
            <span class="n">S</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">S_matrix_id</span><span class="p">]</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">S</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"S_matrix_id '</span><span class="si">{</span><span class="n">S_matrix_id</span><span class="si">}</span><span class="s2">' not found in adata.uns."</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">fig</span></div>


<div class="viewcode-block" id="SCOTCH.combined_enrichment_visualization">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.combined_enrichment_visualization">[docs]</a>
    <span class="k">def</span> <span class="nf">combined_enrichment_visualization</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">enrich_object_id</span><span class="p">,</span> <span class="n">top_k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">max_point_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                                          <span class="n">palette</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span> <span class="n">var1</span><span class="o">=</span><span class="s2">"cell_clusters"</span><span class="p">,</span> <span class="n">var2</span><span class="o">=</span><span class="s2">"sample"</span><span class="p">,</span> <span class="n">S_matrix_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Create a 2x3 subplot visualization combining enrichment bubble plots, element count heatmaps,</span>
<span class="sd">        co-occurrence proportions, and S matrix heatmaps.</span>

<span class="sd">        This visualization provides insights into cellular data enrichment and relationships between variables.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adata : anndata.AnnData</span>
<span class="sd">            An AnnData object containing single-cell data. Must contain `obs`, `uns`, and required data matrices.</span>
<span class="sd">        enrich_object_id : str</span>
<span class="sd">            The identifier for enrichment data in `adata.uns`.</span>
<span class="sd">        top_k : int, optional (default=5)</span>
<span class="sd">            The number of top enrichment terms to display in visualizations.</span>
<span class="sd">            Must be a positive integer.</span>
<span class="sd">        max_point_size : int, optional (default=100)</span>
<span class="sd">            The maximum size for the points in the bubble plot. Determines the largest bubble size.</span>
<span class="sd">        palette : str, optional (default='viridis')</span>
<span class="sd">            The color palette used for the bubble plot.</span>
<span class="sd">        var1 : str, optional</span>
<span class="sd">            The first variable for co-occurrence proportions. Must exist in `adata.obs.columns`.</span>
<span class="sd">        var2 : str, optional</span>
<span class="sd">            The second variable for co-occurrence proportions. Must exist in `adata.obs.columns`.</span>
<span class="sd">        S_matrix_id : str, optional</span>
<span class="sd">            The key in `adata.uns` for the additional heatmap matrix. The associated value must be a 2D matrix.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        matplotlib.figure.Figure</span>
<span class="sd">            A matplotlib figure object containing the generated subplots.</span>
<span class="sd">        """</span>

        <span class="c1"># Validate 'top_k'</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">top_k</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="n">top_k</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"'top_k' must be a positive integer, but got </span><span class="si">{</span><span class="n">top_k</span><span class="si">}</span><span class="s2">."</span><span class="p">)</span>

        <span class="c1"># Validate 'max_point_size'</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">max_point_size</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="n">max_point_size</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"'max_point_size' must be a positive integer, but got </span><span class="si">{</span><span class="n">max_point_size</span><span class="si">}</span><span class="s2">."</span><span class="p">)</span>

        <span class="c1"># Validate 'adata'</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">"obs"</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">"uns"</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"The input `adata` object must have both `obs` and `uns` attributes."</span><span class="p">)</span>

        <span class="c1"># Validate 'var1' and 'var2'</span>
        <span class="k">if</span> <span class="n">var1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">var1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The specified 'var1' ('</span><span class="si">{</span><span class="n">var1</span><span class="si">}</span><span class="s2">') is not a valid column in 'adata.obs'."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">var2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">var2</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The specified 'var2' ('</span><span class="si">{</span><span class="n">var2</span><span class="si">}</span><span class="s2">') is not a valid column in 'adata.obs'."</span><span class="p">)</span>

        <span class="c1"># Validate 'S_matrix_id'</span>
        <span class="k">if</span> <span class="n">S_matrix_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">S_matrix_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"The specified 'S_matrix_id' ('</span><span class="si">{</span><span class="n">S_matrix_id</span><span class="si">}</span><span class="s2">') is not found in 'adata.uns'."</span><span class="p">)</span>
            <span class="n">S</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">S_matrix_id</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">S</span><span class="p">,</span> <span class="s2">"shape"</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">S</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">"The 'S_matrix_id' value must be a 2D matrix, but got an invalid object of type: </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">S</span><span class="p">)</span><span class="si">}</span><span class="s2">."</span><span class="p">)</span>

        <span class="c1"># Setup plot grid:</span>
        <span class="k">if</span> <span class="n">var2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">var2</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">n_set_2</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">var2</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"var2 '</span><span class="si">{</span><span class="n">var2</span><span class="si">}</span><span class="s2">' not found in adata.obs."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">S_matrix_id</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">:</span>
            <span class="n">S</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">S_matrix_id</span><span class="p">]</span>
            <span class="n">gene_factors</span> <span class="o">=</span> <span class="n">S</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">cell_factors</span> <span class="o">=</span> <span class="n">S</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"S_matrix_id '</span><span class="si">{</span><span class="n">S_matrix_id</span><span class="si">}</span><span class="s2">' not found in adata.uns."</span><span class="p">)</span>

        <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
        <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">figure</span><span class="o">=</span><span class="n">fig</span><span class="p">,</span>
                               <span class="n">width_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_set_2</span><span class="p">,</span> <span class="n">gene_factors</span><span class="p">],</span>
                               <span class="n">height_ratios</span><span class="o">=</span><span class="p">[</span><span class="n">cell_factors</span><span class="p">,</span> <span class="n">top_k</span> <span class="o">*</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="n">gene_factors</span><span class="p">])</span>

        <span class="n">ax_A</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>  <span class="c1"># Panel A</span>
        <span class="n">ax_B</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>  <span class="c1"># Panel B</span>
        <span class="n">ax_C</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>  <span class="c1"># Panel C</span>
        <span class="n">ax_D</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>  <span class="c1"># Panel D</span>

        <span class="c1"># Plot 1: Element Count Heatmap</span>
        <span class="k">if</span> <span class="n">var1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plot_element_count_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">var1</span><span class="p">,</span> <span class="s2">"vertical"</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax_A</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"var1 parameter is required for the element count heatmap."</span><span class="p">)</span>
        <span class="n">ax_A</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_A</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_A</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'x'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelbottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labeltop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax_A</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'y'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Plot 2: Co-occurrence Proportions</span>
        <span class="k">if</span> <span class="n">var1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">var2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plot_cooccurrence_proportions</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">var1</span><span class="p">,</span> <span class="n">var2</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax_B</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"Both var1 and var2 are required for co-occurrence proportions."</span><span class="p">)</span>
        <span class="n">ax_B</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_B</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_B</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'x'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelbottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labeltop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax_B</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'y'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Plot S matrix</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plot_S_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span> <span class="n">S_matrix_id</span><span class="o">=</span><span class="n">S_matrix_id</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax_C</span><span class="p">)</span>
        <span class="n">ax_C</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_C</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_C</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'x'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelbottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labeltop</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax_C</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'y'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Plot 4: Enrichment Bubble Plot</span>
        <span class="n">gene_subset</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">S</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">visualize_enrichment_bubbleplots</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">enrich_object_id</span><span class="p">,</span> <span class="n">top_k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">max_point_size</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
                                              <span class="n">gene_cluster_set</span><span class="o">=</span><span class="n">gene_subset</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax_D</span><span class="p">)</span>
        <span class="n">ax_D</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_D</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'x'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelbottom</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">labeltop</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax_D</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'y'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax_D</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">'upper left'</span><span class="p">,</span> <span class="n">borderaxespad</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="c1"># plt.show()</span>
        <span class="k">return</span> <span class="n">fig</span></div>


<div class="viewcode-block" id="SCOTCH.combined_embedding_visualization">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.combined_embedding_visualization">[docs]</a>
    <span class="k">def</span> <span class="nf">combined_embedding_visualization</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span>
                                         <span class="n">gene_cluster_id</span> <span class="o">=</span> <span class="s1">'gene_clusters'</span><span class="p">,</span>
                                         <span class="n">gene_embedding_id</span><span class="o">=</span><span class="s1">'gene_embedding'</span><span class="p">,</span>
                                         <span class="n">top_k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">max_point_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                                         <span class="n">palette</span><span class="o">=</span><span class="s1">'viridis'</span><span class="p">,</span> <span class="n">var1</span><span class="o">=</span><span class="s1">'cell_clusters'</span><span class="p">,</span> <span class="n">var2</span><span class="o">=</span><span class="s1">'sample'</span><span class="p">,</span> <span class="n">S_matrix_id</span><span class="o">=</span><span class="s2">"S_matrix"</span><span class="p">,</span>
                                         <span class="n">prefix</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Generate a combined visualization of embeddings from the data, with options to color by metadata.</span>

<span class="sd">        This function is designed to visualize embeddings (e.g., UMAP, PCA, t-SNE) stored in `adata`, optionally</span>
<span class="sd">        allowing users to color the points by specific metadata columns (like cell type or condition). It arranges</span>
<span class="sd">        one or more plots in a grid layout.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        adata : anndata.AnnData</span>
<span class="sd">            An AnnData object containing single-cell data. Must contain embeddings in `adata.obsm`.</span>
<span class="sd">        gene_embedding_id : str, optional</span>
<span class="sd">            The key in `adata.obsm` where the embedding is stored (e.g., `'X_umap'` for UMAP). Default is `'gene_embedding'`.</span>
<span class="sd">        top_k : int, optional (default=5)</span>
<span class="sd">            the number of top features to display per gene cluster. Must be a positive integer.</span>
<span class="sd">        max_point_size : int, optional (default=2)</span>
<span class="sd">            Max point size for V bubble plot.</span>
<span class="sd">        palette : str, optional (default='viridis')</span>
<span class="sd">            The color palette to use for the scatter plot when coloring points.</span>
<span class="sd">        var1 : str, optional</span>
<span class="sd">            The primary variable (from `adata.obs`) to use for heatmap plotting (e.g., 'cell_clusters').</span>
<span class="sd">        var2 : str, optional</span>
<span class="sd">            The secondary variable (from `adata.obs`) for co-occurrence and proportions (e.g., 'sample').</span>
<span class="sd">        S_matrix_id : str, optional</span>
<span class="sd">            Key in `adata.uns` for an externally referenced matrix (e.g., factor matrix or count matrix).</span>

<span class="sd">         Returns</span>
<span class="sd">        -------</span>
<span class="sd">        matplotlib.figure.Figure</span>
<span class="sd">            A matplotlib figure containing the generated visualizations.</span>

<span class="sd">        """</span>
        <span class="c1"># Setup plot grid:</span>
        <span class="k">if</span> <span class="n">var2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">var2</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">n_set_2</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">var2</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"var2 '</span><span class="si">{</span><span class="n">var2</span><span class="si">}</span><span class="s2">' not found in adata.obs."</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">S_matrix_id</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">:</span>
            <span class="n">S</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">S_matrix_id</span><span class="p">]</span>
            <span class="n">gene_factors</span> <span class="o">=</span> <span class="n">S</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">cell_factors</span> <span class="o">=</span> <span class="n">S</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">"S_matrix_id '</span><span class="si">{</span><span class="n">S_matrix_id</span><span class="si">}</span><span class="s2">' not found in adata.uns."</span><span class="p">)</span>

        <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">12</span><span class="p">))</span>
        <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">figure</span><span class="o">=</span><span class="n">fig</span><span class="p">,</span>
                               <span class="n">width_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_set_2</span><span class="p">,</span> <span class="n">gene_factors</span><span class="p">],</span>
                               <span class="n">height_ratios</span><span class="o">=</span><span class="p">[</span><span class="n">cell_factors</span><span class="p">,</span> <span class="n">top_k</span> <span class="o">*</span> <span class="mf">0.5</span> <span class="o">*</span> <span class="n">gene_factors</span><span class="p">])</span>

        <span class="n">ax_A</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>  <span class="c1"># Panel A</span>
        <span class="n">ax_B</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>  <span class="c1"># Panel B</span>
        <span class="n">ax_C</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>  <span class="c1"># Panel C</span>
        <span class="n">ax_D</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>  <span class="c1"># Panel D</span>

        <span class="c1"># Plot 1: Element Count Heatmap</span>
        <span class="k">if</span> <span class="n">var1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plot_element_count_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">var1</span><span class="p">,</span> <span class="s2">"vertical"</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax_A</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"var1 parameter is required for the element count heatmap."</span><span class="p">)</span>
        <span class="n">ax_A</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_A</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_A</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'x'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelbottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labeltop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax_A</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'y'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Plot 2: Co-occurrence Proportions</span>
        <span class="k">if</span> <span class="n">var1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">var2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plot_cooccurrence_proportions</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">var1</span><span class="p">,</span> <span class="n">var2</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax_B</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">"Both var1 and var2 are required for co-occurrence proportions."</span><span class="p">)</span>
        <span class="n">ax_B</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_B</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_B</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'x'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelbottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labeltop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax_B</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'y'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Plot S matrix</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plot_S_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span> <span class="n">S_matrix_id</span><span class="o">=</span><span class="n">S_matrix_id</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax_C</span><span class="p">)</span>
        <span class="n">ax_C</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_C</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_C</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'x'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelbottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labeltop</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax_C</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'y'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Plot 4: Enrichment Bubble Plot</span>
        <span class="n">gene_subset</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">S</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">visualize_marker_gene_bubbleplot_per_gene_cluster</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">gene_cluster_id</span><span class="o">=</span><span class="n">gene_cluster_id</span><span class="p">,</span>
                                                               <span class="n">gene_embedding_id</span><span class="o">=</span><span class="n">gene_embedding_id</span><span class="p">,</span>
                                                               <span class="n">top_k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">max_point_size</span><span class="o">=</span><span class="n">max_point_size</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax_D</span><span class="p">)</span>
        <span class="n">ax_D</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">ax_D</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'x'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelbottom</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">labeltop</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax_D</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">'y'</span><span class="p">,</span> <span class="n">which</span><span class="o">=</span><span class="s1">'both'</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax_D</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">'upper left'</span><span class="p">,</span> <span class="n">borderaxespad</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="c1"># plt.show()</span>
        <span class="k">return</span> <span class="n">fig</span></div>


    <span class="k">def</span> <span class="nf">_fill_missing_categories</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">categorical_columns</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">):</span>
<span class="w">        </span><span class="sd">"""</span>
<span class="sd">        Fills missing combinations of categories in categorical columns with the specified fill_value.</span>

<span class="sd">        Parameters:</span>
<span class="sd">        - df: pandas DataFrame</span>
<span class="sd">        - categorical_columns: list of columns to consider for filling missing categories</span>
<span class="sd">        - fill_value: value to fill missing data with (default is NaN)</span>

<span class="sd">        Returns:</span>
<span class="sd">        - A DataFrame with all combinations of categories filled.</span>
<span class="sd">        """</span>

        <span class="c1"># Generate all possible combinations of the categorical columns</span>
        <span class="n">all_combinations</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_product</span><span class="p">(</span>
            <span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">categorical_columns</span><span class="p">],</span>
            <span class="n">names</span><span class="o">=</span><span class="n">categorical_columns</span>
        <span class="p">)</span>

        <span class="c1"># Reindex the DataFrame to ensure all combinations are included</span>
        <span class="n">df_filled</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">categorical_columns</span><span class="p">)</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">all_combinations</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="n">fill_value</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">df_filled</span>

<div class="viewcode-block" id="SCOTCH.plot_reconstruction_error">
<a class="viewcode-back" href="../SCOTCH.html#SCOTCH.SCOTCH.plot_reconstruction_error">[docs]</a>
    <span class="k">def</span> <span class="nf">plot_reconstruction_error</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">):</span>

        <span class="c1"># Extract keys matching the pattern '*_reconstruction_error'</span>
        <span class="n">reconstruction_error_keys</span> <span class="o">=</span> <span class="p">[</span><span class="n">key</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span> <span class="k">if</span> <span class="n">key</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">'_reconstruction_error'</span><span class="p">)]</span>

        <span class="c1"># Create line plot for each key</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">reconstruction_error_keys</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">'_reconstruction_error'</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">reconstruction_error</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
            <span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">reconstruction_error</span><span class="p">)),</span> <span class="n">y</span><span class="o">=</span><span class="n">reconstruction_error</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">prefix</span><span class="p">)</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s1">'log'</span><span class="p">)</span>
        <span class="c1">#plt.title('Reconstruction Errors (Log Scale)')</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">'Epoch'</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">'Error'</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">'Regularization'</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">'upper left'</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span></div>
</div>

</pre></div>

          </article>
        </div>
      </div>
    </main>
  </div>
  <footer class="md-footer">
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
          
          
        </a>
        
      </nav>
    </div>
    <div class="md-footer-meta md-typeset">
      <div class="md-footer-meta__inner md-grid">
        <div class="md-footer-copyright">
          <div class="md-footer-copyright__highlight">
              &#169; Copyright 2025, Spencer Halberg-Spencer, Harmon Bhasin, Sushmita Roy.
              
          </div>
            Created using
            <a href="http://www.sphinx-doc.org/">Sphinx</a> 7.3.7.
             and
            <a href="https://github.com/bashtage/sphinx-material/">Material for
              Sphinx</a>
        </div>
      </div>
    </div>
  </footer>
  <script src="../_static/javascripts/application.js"></script>
  <script>app.initialize({version: "1.0.4", url: {base: ".."}})</script>
  </body>
</html>